Documentation to create a distributed applet can be found on the [Developer Tutorials wiki page](https://wiki.dnanexus.com/Developer-Tutorials/Parallelize-Your-App). This readme will focus on the details of this applet.

## How is the SAMtools dependency provided?
The SAMtools dependency is resolved by declaring an [Apt-Get](https://help.ubuntu.com/14.04/serverguide/apt-get.html) package in the `dxapp.json` file's `runSpec.execDepends`.
```json
{
...
    "runSpec": {
  	...
      "execDepends": [
        {"name": "samtools"}
      ]
    }
...
}
```
For additional information, please refer to the [`execDepends`](https://wiki.dnanexus.com/Execution-Environment-Reference#Software-Packages) wiki page.

## Entry Points
Distributed bash-interpreter apps use bash functions to [declare entry points](https://wiki.dnanexus.com/Developer-Tutorials/Parallelize-Your-App#Adding-Entry-Points-to-Your-Code). This app has the following entry points specified as bash functions:

* *main* 
* *count_func*
* *sum_reads*

Entry points are executed on a new worker with its own system requirements. The instance type can be set in the `dxapp.json` file's `runSpec.systemRequirements`:
```json
{
  "runSpec": {
    ...
    "systemRequirements": {
      "main": {
        "instanceType": "mem1_ssd1_x4"
      },
      "count_func": {
        "instanceType": "mem1_ssd1_x2"
      },
      "sum_reads": {
        "instanceType": "mem1_ssd1_x4"
      }
    },
    ...
  }
}
```
## main
The *main* function slices the initial `*.bam` file and generates an index `*.bai` if needed. The input `*.bam` is the sliced into smaller `*.bam` files containing only reads from canonical chromosomes. First, the main function downloads the BAM file and gets the headers.
<!-- SECTION: Downloading inputs and get chromosomes list from headers -->

Sliced `*.bam` files are uploaded and their file IDs are passed to the *count_func* entry point using the [`dx-jobutil-new-job`](https://wiki.dnanexus.com/Helpstrings-of-SDK-Command-Line-Utilities#dx-jobutil-new-job) command.
<!-- SECTION: Split bam into multiple chromosomes and send as input to sam_count subjob -->


Outputs from the *count_func* entry points are referenced as Job Based Object References ([JBOR](https://wiki.dnanexus.com/API-Specification-v1.0.0/Job-Input-and-Output#Job-Dependencies)) and used as inputs for the *sum_reads* entry point.
<!-- SECTION: Gather output of count jobs and write to result file -->

The output of the *sum_reads* entry point is used as the output of the main entry point via JBOR reference using the command [`dx-jobutil-add-output`](https://wiki.dnanexus.com/Helpstrings-of-SDK-Command-Line-Utilities#dx-jobutil-add-output).
<!-- Upload chromosome_results.txt from sum_reads subjob as job output -->

## count_func
This entry point downloads and runs the command `samtools view -c` on the sliced `*.bam`. The generated `counts_txt` output file is uploaded as the entry point's job output via the command [`dx-jobutil-add-output`](https://wiki.dnanexus.com/Helpstrings-of-SDK-Command-Line-Utilities#dx-jobutil-add-output).
<!-- FUNCTION: count_func -->

## sum_reads
The *main* entry point triggers this subjob, providing the output of *count_func* as an input. This entry point gathers all the files generated by the *count_func* jobs and sums them.

This function returns `read_sum_file` as the entry point output.
<!-- FUNCTION: sum_reads -->
