<!DOCTYPE html>

<html lang="en">
	<head>
		<meta charset="utf-8">
		<meta name="viewport" content="width=device-width, initial-scale=1">
		<meta name="generator" content="Jekyll v3.7.4">

		<link rel="stylesheet" href="https://dnanexus.github.io/dnanexus-example-applets/css/screen.css">
		<link rel="apple-touch-icon" href="https://dnanexus.github.io/dnanexus-example-applets/apple-touch-icon.png">
		<link rel="icon" type="image/png" href="https://dnanexus.github.io/dnanexus-example-applets/touch-icon.png" sizes="192x192">
		<link rel="icon" type="image/png" href="https://dnanexus.github.io/dnanexus-example-applets/images/favicon.ico">
		<link rel="stylesheet" href="//fonts.googleapis.com/css?family=Merriweather:400italic,400,300italic,300,700,700italic|Open+Sans:400italic,600italic,700italic,700,600,400|Inconsolata:400,700">
		<script type="text/javascript" src="https://dnanexus.github.io/dnanexus-example-applets/js/jquery.min.js"></script>
		<!-- Begin Jekyll SEO tag v2.5.0 -->
<title>Search | DNAnexus Example Applets</title>
<meta name="generator" content="Jekyll v3.7.4" />
<meta property="og:title" content="Search" />
<meta name="author" content="DNAnexus" />
<meta property="og:locale" content="en_US" />
<meta name="description" content="dnanexus-example-applets github page." />
<meta property="og:description" content="dnanexus-example-applets github page." />
<link rel="canonical" href="https://dnanexus.github.io/dnanexus-example-applets/search/" />
<meta property="og:url" content="https://dnanexus.github.io/dnanexus-example-applets/search/" />
<meta property="og:site_name" content="DNAnexus Example Applets" />
<meta name="twitter:card" content="summary" />
<meta name="twitter:site" content="@dnanexus" />
<meta name="twitter:creator" content="@DNAnexus" />
<script type="application/ld+json">
{"description":"dnanexus-example-applets github page.","publisher":{"@type":"Organization","logo":{"@type":"ImageObject","url":"https://dnanexus.github.io/dnanexus-example-applets/siteicon.png"},"name":"DNAnexus"},"@type":"WebPage","url":"https://dnanexus.github.io/dnanexus-example-applets/search/","headline":"Search","author":{"@type":"Person","name":"DNAnexus"},"@context":"http://schema.org"}</script>
<!-- End Jekyll SEO tag -->

		<link type="application/atom+xml" rel="alternate" href="https://dnanexus.github.io/dnanexus-example-applets/feed.xml" title="DNAnexus Example Applets" />

		
	</head>

	<body class="">
		<header>
			<section class="top-bar">
				<!-- FIXME: Workaround for Jekyll terminal '/' issue -->
				<div class="logo"><a href="https://dnanexus.github.io/dnanexus-example-applets/"><svg
   xmlns:dc="http://purl.org/dc/elements/1.1/"
   xmlns:cc="http://creativecommons.org/ns#"
   xmlns:rdf="http://www.w3.org/1999/02/22-rdf-syntax-ns#"
   xmlns:svg="http://www.w3.org/2000/svg"
   xmlns="http://www.w3.org/2000/svg"
   viewBox="0 0 220.19125 265.54626"
   height="265.54626"
   width="220.19125"
   xml:space="preserve"
   version="1.1"
   id="svg2"><metadata
     id="metadata8"><rdf:RDF><cc:Work
         rdf:about=""><dc:format>image/svg+xml</dc:format><dc:type
           rdf:resource="http://purl.org/dc/dcmitype/StillImage" /></cc:Work></rdf:RDF></metadata><defs
     id="defs6"><clipPath
       id="clipPath16"
       clipPathUnits="userSpaceOnUse"><path
         id="path18"
         d="m 0,212.437 176.153,0 L 176.153,0 0,0 0,212.437 Z" /></clipPath></defs><g
     transform="matrix(1.25,0,0,-1.25,0,265.54625)"
     id="g10"><g
       id="g12"><g
         clip-path="url(#clipPath16)"
         id="g14"><g
           transform="translate(3.2174,209.4236)"
           id="g20"><path
             id="path22"
             style="fill:#97d72d;fill-opacity:1;fill-rule:evenodd;stroke:none"
             d="m 0,0 40.153,0 c 0,-49.386 22.811,-75.017 57.546,-94.788 33.443,-19.03 73.237,-57.663 73.237,-111.344 l -40.153,0 c 0,53.681 -23.74,82.719 -58.314,101.467 C 41.978,-88.132 0,-49.386 0,0" /></g><g
           transform="translate(174.1533,209.437)"
           id="g24"><path
             id="path26"
             style="fill:#0081b0;fill-opacity:1;fill-rule:evenodd;stroke:none"
             d="m 0,0 -40.153,0 c 0,-49.386 -30.658,-70.84 -65.393,-90.608 -33.443,-19.036 -65.39,-61.843 -65.39,-115.524 l 40.153,0 c 0,53.681 30.85,80.988 65.424,99.736 C -34.869,-89.863 0,-49.386 0,0" /></g></g></g></g></svg></a></div>
				<a class="nav-toggle" id="open-nav" href="#">&#9776;</a>
<nav>
	
	
	

		
		<a href="https://dnanexus.github.io/dnanexus-example-applets/tutorials/" class="">Tutorials</a>
	
	
	

		
		<a href="https://dnanexus.github.io/dnanexus-example-applets/resources/" class="">Resources</a>
	
</nav>

			</section>
			<section class="hero_search">
				<h1></h1>
				<p></p>
				<!-- TODO: Workaround to hard link. Find first elem in _set getting-started instead -->
				<div class=btn-wrapper>
				<button type="button" class="btn btn-success btn-lg" onclick="window.location = 'https://dnanexus.github.io/dnanexus-example-applets/overview/'">Get Started</button>
				</div>
				<form action="https://dnanexus.github.io/dnanexus-example-applets/search/" method="get">
	<input type="search" name="q"  placeholder="What would you like to know?" autofocus>
	<svg fill="#000000" height="24" viewBox="0 0 24 24" width="24" xmlns="http://www.w3.org/2000/svg">
    <path d="M15.5 14h-.79l-.28-.27C15.41 12.59 16 11.11 16 9.5 16 5.91 13.09 3 9.5 3S3 5.91 3 9.5 5.91 16 9.5 16c1.61 0 3.09-.59 4.23-1.57l.27.28v.79l5 4.99L20.49 19l-4.99-5zm-6 0C7.01 14 5 11.99 5 9.5S7.01 5 9.5 5 14 7.01 14 9.5 11.99 14 9.5 14z"/>
    <path d="M0 0h24v24H0z" fill="none"/>
</svg>
	<input type="submit" value="Search" style="display: none;">
</form>
			</section>
			</div>

		</header>
		<section class="content">
			<div class="wrapper">
				<p><span id="search-process">Loading</span> results <span id="search-query-container" style="display: none;">for "<strong id="search-query"></strong>"</span></p>
<ul id="search-results"></ul>

<script>
	window.data = {
		
			
				
					
					
					"web-app-bash-tensorboard-web-app": {
						"id": "web-app-bash-tensorboard-web-app",
						"title": "TensorBoard Example Web App ",
						"categories": "web_app, bash",
						"url": " /web_app/bash/tensorboard-web-app/",
						"content": "This example demonstrates how to run TensorBoard inside a DNAnexus applet.\n\nTensorBoard is a web application used to visualize and inspect what is going on inside TensorFlow training. To use TensorBoard, our training script in TensorFlow needs to include code that saves various data to a log directory where TensorBoard can then find the data to display it.\n\nThis example uses an example script from the TensorBoard authors. For more guidance on how to use TensorBoard, check out the tensorflow website (external link).\n\nCreating the web application\nThe applet code runs a training script, which is placed in resources/home/dnanexus/ to make it available in the current working directory of the worker, and then it starts tensorboard on port 443 (HTTPS).\n# Start the training script and put it into the background,\n# so the next line of code will run immediately\npython mnist_tensorboard_example.py --log_dir LOGS_FOR_TENSORBOARD &amp;\n\n# Run TensorBoard\ntensorboard  --logdir LOGS_FOR_TENSORBOARD --host 0.0.0.0 --port 443\n\nWe run the training script in the background to start TensorBoard immediately, which will let us see the results while training is still running. This is particularly important for long-running training scripts.\n\nNote that for all web apps, if everything is running smoothly and no errors are encountered (the ideal case), the line of code that starts the server will keep it running forever. The applet stops only when it is terminated. This also means that any lines of code after the server starts will not be executed.\n\nAs with all web apps, the dxapp.json must include \"httpsApp\": {\"ports\":[443], \"shared_access\": \"VIEW\"} to tell the worker to expose port 443.\n\nCreating an applet on DNAnexus\nBuild the asset with the libraries first:\ndx build_asset tensorflow_asset\n\n\nTake the record ID it outputs and add it to the dxapp.json for the applet.\n\"runSpec\": {\n\t...\n\t\"assetDepends\": [\n    {\n      \"id\": \"record-xxxx\n    }\n  ]\n\t...\n}\n\n\nThen build the applet\ndx build -f tensorboard-web-app\ndx run tensorboard-web-app\n\nOnce it spins up, you can go to that job’s designated URL based on its job ID, https://job-xxxx.dnanexus.cloud/, to see the result."
					}
					
				
			
		
			
				
					,
					
					"python-parallel-samtools-count-para-reg-multiprocess-py": {
						"id": "python-parallel-samtools-count-para-reg-multiprocess-py",
						"title": "Parallel by Region (py)",
						"categories": "python, parallel",
						"url": " /python/parallel/samtools_count_para_reg_multiprocess_py/",
						"content": "This applet tutorial will perform a SAMtools count using parallel threads.\n\nIn order to take full advantage of the scalability that cloud computing offers, our scripts have to implement the correct methodologies. This applet tutorial will:\n\n  Install SAMtools\n  Download BAM file\n  Split workload\n  Count regions in parallel\n\n\nThis applet tutorial code is extremely similar to the Parallel Threads SAMtools count tutorial, except multiprocessing is used instead of multiprocessing.dummy.\n\nHow is the SAMtools dependency provided?\nThe SAMtools dependency is resolved by declaring an Apt-Get package in the dxapp.json runSpec.execDepends field.\n{\n  \"runSpec\": {\n    ...\n    \"execDepends\": [\n      {\"name\": \"samtools\"}\n    ]\n  }\n\nFor additional information, please refer to the execDepends wiki page.\n\nDownload Inputs\n\nThis applet downloads all inputs at once using dxpy.download_all_inputs:\ninputs = dxpy.download_all_inputs()\n# download_all_inputs returns a dictionary that contains mapping from inputs to file locations.\n# Additionaly, helper keys, value pairs are added to the dicitonary, similar to bash helper functions\ninputs\n#     mappings_sorted_bam_path: [u'/home/dnanexus/in/mappings_sorted_bam/SRR504516.bam']\n#     mappings_sorted_bam_name: u'SRR504516.bam'\n#     mappings_sorted_bam_prefix: u'SRR504516'\n#     mappings_sorted_bai_path: u'/home/dnanexus/in/mappings_sorted_bai/SRR504516.bam.bai'\n#     mappings_sorted_bai_name: u'SRR504516.bam.bai'\n#     mappings_sorted_bai_prefix: u'SRR504516'\n\nSplit workload\nWe parallely process by using the python multiprocessing module. A rather simple pattern you can follow to compute in a parallel manner is shown below:\nprint \"Number of cpus: {0}\".format(cpu_count())  # Get cpu count from multiprocessing\nworker_pool = Pool(processes=cpu_count())  # Create a pool of workers, 1 for each core\nresults = worker_pool.map(run_cmd, view_cmds)  # map run_cmds to a collection\n                                               # Pool.map will handle orchestrating the job\nworker_pool.close()\nworker_pool.join()  # Make sure to close and join workers when done\n\nThis convenient pattern allows you to quickly orchestrate jobs on a worker. For more detailed overview of the multiprocessing module, visit the python2.7 docs.\n\nWe create several helpers in our applet script to manage our workload. One helper you may have seen before is run_cmd; we use this function to manage or subprocess calls:\ndef run_cmd(cmd_arr):\n    \"\"\"Run shell command.\n    Helper function to simplify the pool.map() call in our parallelization.\n    Raises OSError if command specified (index 0 in cmd_arr) isn't valid\n    \"\"\"\n    proc = subprocess.Popen(\n        cmd_arr,\n        stdout=subprocess.PIPE,\n        stderr=subprocess.PIPE)\n    stdout, stderr = proc.communicate()\n    exit_code = proc.returncode\n    proc_tuple = (stdout, stderr, exit_code)\n    return proc_tuple\n\n\nBefore we can split our workload, we need to know what regions are present in our BAM input file. We handle this initial parsing in the parse_sam_header_for_region function:\ndef parse_sam_header_for_region(bamfile_path):\n    \"\"\"Helper function to match SN regions contained in SAM header\n\n    Returns:\n        regions (list[string]): list of regions in bam header\n    \"\"\"\n    header_cmd = ['samtools', 'view', '-H', bamfile_path]\n    print 'parsing SAM headers:', \" \".join(header_cmd)\n    headers_str = subprocess.check_output(header_cmd)\n    rgx = re.compile(r'SN:(\\S+)\\s')\n    regions = rgx.findall(headers_str)\n    return regions\n\n\nOnce our workload is split and we’ve started processing, we wait and review the status of each Pool worker. Then, we merge and output our results.\n# Write results to file\nresultfn = inputs['mappings_sorted_bam_name'][0]\nresultfn = (\n    resultfn[:-4] + '_count.txt'\n    if resultfn.endswith(\".bam\")\n    else resultfn + '_count.txt')\nwith open(resultfn, 'w') as f:\n    sum_reads = 0\n    for res, reg in izip(results, regions):\n        read_count = int(res[0])\n        sum_reads += read_count\n        f.write(\"Region {0}: {1}\\n\".format(reg, read_count))\n    f.write(\"Total reads: {0}\".format(sum_reads))\n\ncount_file = dxpy.upload_local_file(resultfn)\noutput = {}\noutput[\"count_file\"] = dxpy.dxlink(count_file)\nreturn output\n\n\n Note: The run_cmd function returns a tuple containing the stdout, stderr, and exit code of the subprocess call. We parse these outputs from our workers to determine whether the run failed or passed.\ndef verify_pool_status(proc_tuples):\n    \"\"\"\n    Helper to verify worker succeeded.\n\n    As failed commands are detected we write the stderr from that command\n    to the job_error.json file. This file will be printed to the platform\n    job log on App failure.\n    \"\"\"\n    all_succeed = True\n    err_msgs = []\n    for proc in proc_tuples:\n        if proc[2] != 0:\n            all_succeed = False\n            err_msgs.append(proc[1])\n    if err_msgs:\n        raise dxpy.exceptions.AppInternalError(\"\\n\".join(err_msgs))"
					}
					
				
			
		
			
				
					,
					
					"parallel-bash-samtools-count-para-chr-xargs-sh": {
						"id": "parallel-bash-samtools-count-para-chr-xargs-sh",
						"title": "Parallel xargs by Chr",
						"categories": "parallel, bash",
						"url": " /parallel/bash/samtools_count_para_chr_xargs_sh/",
						"content": "This applet slices a BAM file by canonical chromosome then performs a parallelized samtools view -c using xargs. Type man xargs for general usage information.\n\nHow is the SAMtools dependency provided?\nThe SAMtools compiled binary is placed directory in the &lt;applet dir&gt;/resources directory. Any files found in the resources/ directory will be uploaded so that they will be present in the root directory of the worker. In our case:\n├── Applet dir\n│   ├── src\n│   ├── dxapp.json\n│   ├── resources\n│       ├── usr\n│           ├── bin\n│               ├── &lt; samtools binary &gt;\n\nWhen this applet is run on a worker, the resources/ folder will be placed in the worker’s root directory /:\n/\n├── usr\n│   ├── bin\n│       ├── &lt; samtools binary &gt;\n├── home\n│   ├── dnanexus\n\n/usr/bin is part of the $PATH variable, so in our script, we can reference the samtools command directly, as in samtools view -c ...\n\nParallel Run\nSplice BAM\nFirst, we download our BAM file and slice it by canonical chromosome, writing the *bam file names to another file.\n\nIn order to split a BAM by regions, we need to have a *.bai index. You can either create an app(let) which takes the *.bai as an input or generate a *.bai in the applet. In this tutorial, we generate the *.bai in the applet, sorting the BAM if necessary.\n\n  dx download \"${mappings_bam}\"\n\n  indexsuccess=true\n  bam_filename=\"${mappings_bam_name}\"\n  samtools index \"${mappings_bam_name}\" || indexsuccess=false\n  if [[ $indexsuccess == false ]]; then\n    samtools sort -o \"${mappings_bam_name}\" \"${mappings_bam_name}\"\n    samtools index \"${mappings_bam_name}\"\n    bam_filename=\"${mappings_bam_name}\"\n  fi\n\n\n  chromosomes=$(samtools view -H \"${bam_filename}\" | grep \"\\@SQ\" | awk -F '\\t' '{print $2}' | awk -F ':' '{if ($2 ~ /^chr[0-9XYM]+$|^[0-9XYM]/) {print $2}}')\n\n  for chr in $chromosomes; do\n    samtools view -b \"${bam_filename}\" \"${chr}\" -o \"bam_${chr}.\"bam\n    echo \"bam_${chr}.bam\"\n  done &gt; bamfiles.txt\n\n\nXargs SAMtools view\nIn the previous section, we recorded the name of each sliced BAM file into a record file. Now we will perform a samtools view -c on each slice using the record file as input.\n  counts_txt_name=\"${mappings_bam_prefix}_count.txt\"\n\n  sum_reads=$(&lt;bamfiles.txt xargs -I {} samtools view -c $view_options '{}' | awk '{s+=$1} END {print s}')\n  echo \"Total Count: ${sum_reads}\" &gt; \"${counts_txt_name}\"\n\n\nUpload results\nThe results file is uploaded using the standard bash process:\n\n  Upload a file to the job execution’s container.\n  Provide the DNAnexus link as a job’s output using the script dx-jobutil-add-output &lt;output name&gt;\n      counts_txt_id=$(dx upload \"${counts_txt_name}\" --brief)\n  dx-jobutil-add-output counts_txt \"${counts_txt_id}\" --class=file"
					}
					
				
			
		
			
				
					,
					
					"python-parallel-samtools-count-para-chr-subprocess-py": {
						"id": "python-parallel-samtools-count-para-chr-subprocess-py",
						"title": "Parallel by Chr (py)",
						"categories": "python, parallel",
						"url": " /python/parallel/samtools_count_para_chr_subprocess_py/",
						"content": "This applet tutorial will perform a SAMtools count using parallel threads.\n\nIn order to take full advantage of the scalability that cloud computing offers, our scripts have to implement the correct methodologies. This applet tutorial will:\n\n  Install SAMtools\n  Download BAM file\n  Count regions in parallel\n\n\nThis applet tutorial code is similar to the Parallel Cores SAMtools count tutorial, except multiprocessing.dummy is used instead of multiprocessing.\n\nHow is the SAMtools dependency provided?\nThe SAMtools dependency is resolved by declaring an Apt-Get package in the dxapp.json runSpec.execDepends field.\n  \"runSpec\": {\n    ...\n    \"execDepends\": [\n      {\"name\": \"samtools\"}\n    ]\n  }\n\nFor additional information, please refer to the execDepends wiki page.\n\nDownload BAM file\n\nThe dxpy.download_all_inputs() function downloads all input files into the /home/dnanexus/in directory. A folder will be created for each input and the file(s) will be downloaded to that directory. For convenience, the dxpy.download_all_inputs function returns a dictionary containing the following keys:\n\n  &lt;var&gt;_path (string): full absolute path to where the file was downloaded.\n  &lt;var&gt;_name (string): name of the file, including extention.\n  &lt;var&gt;_prefix (string): name of the file minus the longest matching pattern found in the dxapp.json I/O pattern field.\n\n\nThe path, name, and prefix key-value pattern is repeated for all applet file class inputs specified in the dxapp.json. In this example, our dictionary has the following key-value pairs:\n{\n    mappings_bam_path: [u'/home/dnanexus/in/mappings_bam/SRR504516.bam']\n    mappings_bam_name: [u'SRR504516.bam']\n    mappings_bam_prefix: [u'SRR504516']\n    index_file_path: [u'/home/dnanexus/in/index_file/SRR504516.bam.bai']\n    index_file_name: [u'SRR504516.bam.bai']\n    index_file_prefix: [u'SRR504516']\n}\n\n    inputs = dxpy.download_all_inputs()\n    shutil.move(inputs['mappings_bam_path'][0], os.getcwd())\n\n\nCount Regions in Parallel\nBefore we can perform our parallel SAMtools count, we must determine the workload for each thread. We arbitrarily set our number of workers to 10 and set the workload per thread to 1 chromosome at a time. There are various ways to achieve multithreaded processing in python. For the sake of simplicity, we use multiprocessing.dummy, a wrapper around Python’s threading module.\n\n    input_bam = inputs['mappings_bam_name'][0]\n\n    bam_to_use = create_index_file(input_bam)\n    print \"Dir info:\"\n    print os.listdir(os.getcwd())\n\n    regions = parseSAM_header_for_region(bam_to_use)\n\n    view_cmds = [\n        create_region_view_cmd(bam_to_use, region)\n        for region\n        in regions]\n\n    print 'Parallel counts'\n    t_pools = ThreadPool(10)\n    results = t_pools.map(run_cmd, view_cmds)\n    t_pools.close()\n    t_pools.join()\n\n    verify_pool_status(results)\n\nEach worker creates a string to be called in a subprocess.Popen call. We use the multiprocessing.dummy.Pool.map(&lt;func&gt;, &lt;iterable&gt;) function to call the helper function run_cmd for each string in the iterable of view commands. Because we perform our multithreaded processing using subprocess.Popen, we will not be alerted to any failed processes. We verify our closed workers in the verify_pool_status helper function.\ndef verify_pool_status(proc_tuples):\n    err_msgs = []\n    for proc in proc_tuples:\n        if proc[2] != 0:\n            err_msgs.append(proc[1])\n    if err_msgs:\n        raise dxpy.exceptions.AppInternalError(\"\\n\".join(err_msgs))\n\n\n Important: In this example we use subprocess.Popen to process and verify our results in verify_pool_status. In general, it is considered good practice to use python’s built-in subprocess convenience functions. In this case, subprocess.check_call would achieve the same goal.\n\nGather Results\n\nEach worker returns a read count of just one region in the BAM file. We sum and output the results as the job output. We use the dx-toolkit python SDK’s dxpy.upload_local_file function to upload and generate a DXFile corresponding to our result file.\n\nFor python, job outputs have to be a dictionary of key-value pairs, with the keys being job output names as defined in the dxapp.json and the values being the output values for corresponding output classes. For files, the output type is a DXLink. We use the dxpy.dxlink function to generate the appropriate DXLink value.\n    resultfn = bam_to_use[:-4] + '_count.txt'\n    with open(resultfn, 'w') as f:\n        sum_reads = 0\n        for res, reg in izip(results, regions):\n            read_count = int(res[0])\n            sum_reads += read_count\n            f.write(\"Region {0}: {1}\\n\".format(reg, read_count))\n        f.write(\"Total reads: {0}\".format(sum_reads))\n\n    count_file = dxpy.upload_local_file(resultfn)\n    output = {}\n    output[\"count_file\"] = dxpy.dxlink(count_file)\n\n    return output"
					}
					
				
			
		
			
				
					,
					
					"parallel-bash-samtools-count-para-chr-busyproc-sh": {
						"id": "parallel-bash-samtools-count-para-chr-busyproc-sh",
						"title": "Parallel by Region (sh)",
						"categories": "parallel, bash",
						"url": " /parallel/bash/samtools_count_para_chr_busyproc_sh/",
						"content": "This applet performs a basic SAMtools count on a series of sliced (by canonical chromosome) BAM files in parallel using wait (Ubuntu 14.04+).\n\nHow is the SAMtools dependency provided?\nThe SAMtools dependency is resolved by declaring an Apt-Get package in the dxapp.json runSpec.execDepends.\n  \"runSpec\": {\n    ...\n    \"execDepends\": [\n      {\"name\": \"samtools\"}\n    ]\n  }\n\nFor additional information, please refer to the execDepends wiki page.\n\nDebugging\nThe command set -e -x -o pipefail will assist you in debugging this applet:\n\n  -e causes the shell to immediately exit if a command returns a non-zero exit code.\n  -x prints commands as they are executed, which is very useful for tracking the job’s status or pinpointing the exact execution failure.\n  -o pipefail makes the return code the first non-zero exit code. (Typically, the return code of pipes is the exit code of the last command, which can create difficult to debug problems.)\n    set -e -x -o pipefail\necho \"Value of mappings_sorted_bam: '${mappings_sorted_bam}'\"\necho \"Value of mappings_sorted_bai: '${mappings_sorted_bai}'\"\n\nmkdir workspace\ncd workspace\ndx download \"${mappings_sorted_bam}\"\n\nif [ -z \"$mappings_sorted_bai\" ]; then\n  samtools index \"$mappings_sorted_bam_name\"\nelse\n  dx download \"${mappings_sorted_bai}\"\nfi\n    \n    The *.bai file was an optional job input. We check for a empty or unset var using the bash built-in test [[ - z ${var}} ]]. Then, we can download or create a *.bai index as needed.\n  \n\n\nParallel Run\nBash’s job control system allows for easy management of multiple processes. In this example, we run bash commands in the background as we control maximum job executions in the foreground.\nWe place processes in the background using the character &amp; after a command.\n  chromosomes=$(samtools view -H \"${mappings_sorted_bam_name}\" | grep \"\\@SQ\" | awk -F '\\t' '{print $2}' | awk -F ':' '{if ($2 ~ /^chr[0-9XYM]+$|^[0-9XYM]/) {print $2}}')\n\n  for chr in $chromosomes; do\n    samtools view -b \"${mappings_sorted_bam_name}\" \"${chr}\" -o \"bam_${chr}.bam\"\n    echo \"bam_${chr}.bam\"\n  done &gt; bamfiles.txt\n\n  busyproc=0\n  while read -r b_file; do\n    echo \"${b_file}\"\n    if [[ \"${busyproc}\" -ge \"$(nproc)\" ]]; then\n      echo Processes hit max\n      while [[ \"${busyproc}\" -gt  0 ]]; do\n        wait -n # p_id\n        busyproc=$((busyproc-1))\n      done\n    fi\n    samtools view -c \"${b_file}\"&gt; \"count_${b_file%.bam}\" &amp;\n    busyproc=$((busyproc+1))\n  done &lt;bamfiles.txt\n\n  while [[ \"${busyproc}\" -gt  0 ]]; do\n    wait -n # p_id\n    busyproc=$((busyproc-1))\n  done\n\nJob Output\nOnce the input bam has been sliced, counted, and summed, the output counts_txt is uploaded using the command dx-upload-all-outputs. The following directory structure required for dx-upload-all-outputs is below:\n├── $HOME\n│   ├── out\n│       ├── &lt; output name in dxapp.json &gt;\n│           ├── output file\n\n\nIn our applet, we upload all outputs by:\n  outputdir=\"${HOME}/out/counts_txt\"\n  mkdir -p \"${outputdir}\"\n  cat count* | awk '{sum+=$1} END{print \"Total reads = \",sum}' &gt; \"${outputdir}/${mappings_sorted_bam_prefix}_count.txt\"\n\n  dx-upload-all-outputs"
					}
					
				
			
		
			
				
					,
					
					"bash-samtools-count-git-sh": {
						"id": "bash-samtools-count-git-sh",
						"title": "Git Dependency",
						"categories": "bash",
						"url": " /bash/samtools_count_git_sh/",
						"content": "What does this applet do?\n\nThis applet performs a basic SAMtools count of alignments present in an input BAM.\n\nPrerequisites\n\nThe app must have network access to the hostname where the git repository is located. In this example, access.network is set to:\n\"access\": {\n  \"network\": [\"github.com\"]\n}\n\nAdditional documentation on access and network fields can be found on the Execution Environment Reference wiki page.\n\nHow is the SAMtools dependency added?\n\nSAMtools is cloned and built from the SAMtools GitHub repository. Let’s take a closer look at the dxapp.json file’s runSpec.execDepends property:\n  \"runSpec\": {\n ...\n    \"execDepends\": [\n        {\n        \"name\": \"htslib\",\n        \"package_manager\": \"git\",\n        \"url\": \"https://github.com/samtools/htslib.git\",\n        \"tag\": \"1.3.1\",\n        \"destdir\": \"/home/dnanexus\"\n        },\n        {\"name\": \"samtools\",\n        \"package_manager\": \"git\",\n        \"url\": \"https://github.com/samtools/samtools.git\",\n        \"tag\": \"1.3.1\",\n        \"destdir\": \"/home/dnanexus\",\n        \"build_commands\": \"make samtools\"\n        }\n    ],\n...\n  }\n\nThe execDepends value is a JSON array of dependencies to resolve before the applet source code is run. In this applet, we specify the following git fetch dependency for htslib and SAMtools. Dependencies are resolved in the order they’re specified. Here we must specify htslib first, before samtools build_commands, due to newer versions of SAMtools depending on htslib. An overview of the each property in the git dependency:\n\n\n  package_manager - Details the type of dependency and how to resolve.  supplementary details.\n  url - Must point to the server containing the repository. In this case, a github url.\n  tag/branch - Git tag/branch to fetch.\n  destdir - Directory on worker to which the git repo is cloned.\n  build_commands - If needed, build commands to execute. We know our first dependency, htslib, is built when we build SAMtools; as a result, we only specify “build_commands” for the SAMtools dependency.\n\n\n Note: build_commands are executed from the destdir; use cd when appropriate.\n\nHow is SAMtools called in our src script?\n\nBecause we set \"destdir\": \"/home/dnanexus\" in our dxapp.json, we know the git repo is cloned to the same directory from which our script will execute. Our example directory’s structure:\n├── home\n│   ├── dnanexus\n│       ├── &lt; app script &gt;\n│       ├── htslib\n│       ├── samtools\n│           ├── &lt; samtools binary &gt;\n\nOur samtools command from the app script is samtools/samtools.\n\nApplet Script\nmain() {\n  set -e -x -o pipefail\n\n  dx download \"$mappings_bam\"\n\n  count_filename=\"${mappings_bam_prefix}.txt\"\n  readcount=$(samtools/samtools view -c \"${mappings_bam_name}\")\n  echo \"Total reads: ${readcount}\" &gt; \"${count_filename}\"\n\n  counts_txt=$(dx upload \"${count_filename}\" --brief)\n  dx-jobutil-add-output counts_txt \"${counts_txt}\" --class=file\n}\n\n\n Note: We could’ve built samtools in a destination within our $PATH or added the binary directory to our $PATH. Keep this in mind for your app(let) development"
					}
					
				
			
		
			
				
					,
					
					"python-distributed-samtools-count-distr-region-py": {
						"id": "python-distributed-samtools-count-distr-region-py",
						"title": "Distributed by Region (py)",
						"categories": "python, distributed",
						"url": " /python/distributed/samtools_count_distr_region_py/",
						"content": "The applet will create a count of reads from a BAM format file. Documentation to create a distributed applet can be found on the DNAnexus wiki. This readme will focus on the details of this applet.\n\nHow is the SAMtools dependency provided?\nThe SAMtools dependency is resolved by declaring an Apt-Get package in the dxapp.json runSpec.execDepends.\n  \"runSpec\": {\n    ...\n    \"execDepends\": [\n      {\"name\": \"samtools\"}\n    ]\n  }\n\nFor additional information, please refer to the execDepends wiki page.\n\nEntry Points\nDistributed python-interpreter apps use python decorators on functions to declare entry points. This app has the following entry points as decorated functions:\n\n\n  main\n  samtoolscount_bam\n  combine_files\n\n\nEntry points are executed on a new worker with their own system requirements. In this example, we split and merge our files on basic mem1_ssd1_x2 instances and perform our own, more intensive, processing step on a mem1_ssd1_x4 instance. Instance type can be set in the dxapp.json runSpec.systemRequirements:\n  \"runSpec\": {\n    ...\n    \"systemRequirements\": {\n      \"main\": {\n        \"instanceType\": \"mem1_ssd1_x2\"\n      },\n      \"samtoolscount_bam\": {\n        \"instanceType\": \"mem1_ssd1_x4\"\n      },\n      \"combine_files\": {\n        \"instanceType\": \"mem1_ssd1_x2\"\n      }\n    },\n    ...\n  }\n\nmain\nThe main function scatters by region bins based on user input. If no *.bai file is present, the applet generates an index *.bai.\n    regions = parseSAM_header_for_region(filename)\n    split_regions = [regions[i:i + region_size]\n                     for i in xrange(0, len(regions), region_size)]\n\n    if not index_file:\n        mappings_bam, index_file = create_index_file(filename, mappings_bam)\n\nRegions bins are passed to the samtoolscount_bam entry point using the dxpy.new_dxjob function.\n    print 'creating subjobs'\n    subjobs = [dxpy.new_dxjob(\n               fn_input={\"region_list\": split,\n                         \"mappings_bam\": mappings_bam,\n                         \"index_file\": index_file},\n               fn_name=\"samtoolscount_bam\")\n               for split in split_regions]\n\n    fileDXLinks = [subjob.get_output_ref(\"readcount_fileDX\")\n                   for subjob in subjobs]\n\nOutputs from the samtoolscount_bam entry points are used as inputs for the combine_files entry point. The output of the combine_files entry point is used as the output of the main entry point.\n    print 'combining outputs'\n    postprocess_job = dxpy.new_dxjob(\n        fn_input={\"countDXlinks\": fileDXLinks, \"resultfn\": filename},\n        fn_name=\"combine_files\")\n\n    countDXLink = postprocess_job.get_output_ref(\"countDXLink\")\n\n    output = {}\n    output[\"count_file\"] = countDXLink\n\n    return output\n\n\nsamtoolscount_bam\nThis entry point downloads and creates a samtools view -c command for each region in the input bin. The dictionary returned from dxpy.download_all_inputs() is used to reference input names and paths.\ndef samtoolscount_bam(region_list, mappings_bam, index_file):\n    \"\"\"Processing function.\n\n    Arguments:\n        region_list (list[str]): Regions to count in BAM\n        mappings_bam (dict): dxlink to input BAM\n        index_file (dict): dxlink to input BAM\n\n    Returns:\n        Dictionary containing dxlinks to the uploaded read counts file\n    \"\"\"\n    #\n    # Download inputs\n    # -------------------------------------------------------------------\n    # dxpy.download_all_inputs will download all input files into\n    # the /home/dnanexus/in directory.  A folder will be created for each\n    # input and the file(s) will be download to that directory.\n    #\n    # In this example our dictionary inputs has the following key, value pairs\n    # Note that the values are all list\n    #     mappings_bam_path: [u'/home/dnanexus/in/mappings_bam/&lt;bam filename&gt;.bam']\n    #     mappings_bam_name: [u'&lt;bam filename&gt;.bam']\n    #     mappings_bam_prefix: [u'&lt;bam filename&gt;']\n    #     index_file_path: [u'/home/dnanexus/in/index_file/&lt;bam filename&gt;.bam.bai']\n    #     index_file_name: [u'&lt;bam filename&gt;.bam.bai']\n    #     index_file_prefix: [u'&lt;bam filename&gt;']\n    #\n\n    inputs = dxpy.download_all_inputs()\n\n    # SAMtools view command requires the bam and index file to be in the same\n    shutil.move(inputs['mappings_bam_path'][0], os.getcwd())\n    shutil.move(inputs['index_file_path'][0], os.getcwd())\n    input_bam = inputs['mappings_bam_name'][0]\n\n    #\n    # Per region perform SAMtools count.\n    # --------------------------------------------------------------\n    # Output count for regions and return DXLink as job output to\n    # allow other entry points to download job output.\n    #\n\n    with open('read_count_regions.txt', 'w') as f:\n        for region in region_list:\n                view_cmd = create_region_view_cmd(input_bam, region)\n                region_proc_result = run_cmd(view_cmd)\n                region_count = int(region_proc_result[0])\n                f.write(\"Region {0}: {1}\\n\".format(region, region_count))\n    readcountDXFile = dxpy.upload_local_file(\"read_count_regions.txt\")\n    readCountDXlink = dxpy.dxlink(readcountDXFile.get_id())\n\n    return {\"readcount_fileDX\": readCountDXlink}\n\nThis entry point returns {\"readcount_fileDX\": readCountDXlink}, a JBOR referencing an uploaded text file. This approach to scatter-gather stores the results in files and uploads/downloads the information as needed. This approach exaggerates a scatter-gather for tutorial purposes. You’re able to pass types other than file such as int.\ncombine_files\nThe main entry point triggers this subjob, providing the output of samtoolscount_bam as an input. This entry point gathers all the files generated by the samtoolscount_bam jobs and sums them.\ndef combine_files(countDXlinks, resultfn):\n    \"\"\"The 'gather' subjob of the applet.\n\n    Arguments:\n        countDXlinks (list[dict]): list of DXlinks to process job output files.\n        resultfn (str): Filename to use for job output file.\n\n    Returns:\n        DXLink for the main function to return as the job output.\n\n    Note: Only the DXLinks are passed as parameters.\n    Subjobs work on a fresh instance so files must be downloaded to the machine\n    \"\"\"\n    if resultfn.endswith(\".bam\"):\n        resultfn = resultfn[:-4] + '.txt'\n\n    sum_reads = 0\n    with open(resultfn, 'w') as f:\n        for i, dxlink in enumerate(countDXlinks):\n            dxfile = dxpy.DXFile(dxlink)\n            filename = \"countfile{0}\".format(i)\n            dxpy.download_dxfile(dxfile, filename)\n            with open(filename, 'r') as fsub:\n                for line in fsub:\n                    sum_reads += parse_line_for_readcount(line)\n                    f.write(line)\n        f.write('Total Reads: {0}'.format(sum_reads))\n\n    countDXFile = dxpy.upload_local_file(resultfn)\n    countDXlink = dxpy.dxlink(countDXFile.get_id())\n\n    return {\"countDXLink\": countDXlink}\n\n\n Important: While the main entry point triggers the processing and gathering entry points, keep in mind the main entry point doesn’t do any heavy lifting or processing. Notice in the .runSpec json above we start with a lightwieght instance, scale up for the processing entry point, then finally scale down for the gathering step."
					}
					
				
			
		
			
				
					,
					
					"distributed-bash-samtools-count-distr-chr-slice-sh": {
						"id": "distributed-bash-samtools-count-distr-chr-slice-sh",
						"title": "Distributed by Chr (sh)",
						"categories": "distributed, bash",
						"url": " /distributed/bash/samtools_count_distr_chr_slice_sh/",
						"content": "Documentation to create a distributed applet can be found on the Developer Tutorials wiki page. This readme will focus on the details of this applet.\n\nHow is the SAMtools dependency provided?\nThe SAMtools dependency is resolved by declaring an Apt-Get package in the dxapp.json file’s runSpec.execDepends.\n{\n...\n    \"runSpec\": {\n  \t...\n      \"execDepends\": [\n        {\"name\": \"samtools\"}\n      ]\n    }\n...\n}\n\nFor additional information, please refer to the execDepends wiki page.\n\nEntry Points\nDistributed bash-interpreter apps use bash functions to declare entry points. This app has the following entry points specified as bash functions:\n\n\n  main\n  count_func\n  sum_reads\n\n\nEntry points are executed on a new worker with its own system requirements. The instance type can be set in the dxapp.json file’s runSpec.systemRequirements:\n{\n  \"runSpec\": {\n    ...\n    \"systemRequirements\": {\n      \"main\": {\n        \"instanceType\": \"mem1_ssd1_x4\"\n      },\n      \"count_func\": {\n        \"instanceType\": \"mem1_ssd1_x2\"\n      },\n      \"sum_reads\": {\n        \"instanceType\": \"mem1_ssd1_x4\"\n      }\n    },\n    ...\n  }\n}\n\nmain\nThe main function slices the initial *.bam file and generates an index *.bai if needed. The input *.bam is the sliced into smaller *.bam files containing only reads from canonical chromosomes. First, the main function downloads the BAM file and gets the headers.\n  dx download \"${mappings_sorted_bam}\"\n  chromosomes=$(samtools view -H \"${mappings_sorted_bam_name}\" | grep \"\\@SQ\" | awk -F '\\t' '{print $2}' | awk -F ':' '{if ($2 ~ /^chr[0-9XYM]+$|^[0-9XYM]/) {print $2}}')\n\n\nSliced *.bam files are uploaded and their file IDs are passed to the count_func entry point using the dx-jobutil-new-job command.\n  if [ -z \"${mappings_sorted_bai}\" ]; then\n    samtools index \"${mappings_sorted_bam_name}\"\n  else\n    dx download \"${mappings_sorted_bai}\" -o \"${mappings_sorted_bam_name}\".bai\n  fi\n\n  count_jobs=()\n  for chr in $chromosomes; do\n    seg_name=\"${mappings_sorted_bam_prefix}_${chr}\".bam\n    samtools view -b \"${mappings_sorted_bam_name}\" \"${chr}\" &gt; \"${seg_name}\"\n    bam_seg_file=$(dx upload \"${seg_name}\" --brief)\n    count_jobs+=($(dx-jobutil-new-job -isegmentedbam_file=\"${bam_seg_file}\" -ichr=\"${chr}\" count_func))\n  done\n\n\nOutputs from the count_func entry points are referenced as Job Based Object References (JBOR) and used as inputs for the sum_reads entry point.\n  for job in \"${count_jobs[@]}\"; do\n    readfiles+=(\"-ireadfiles=${job}:counts_txt\")\n  done\n\n  sum_reads_job=$(dx-jobutil-new-job \"${readfiles[@]}\" -ifilename=\"${mappings_sorted_bam_prefix}\" sum_reads)\n\n\nThe output of the sum_reads entry point is used as the output of the main entry point via JBOR reference using the command dx-jobutil-add-output.\n\n\ncount_func\nThis entry point downloads and runs the command samtools view -c on the sliced *.bam. The generated counts_txt output file is uploaded as the entry point’s job output via the command dx-jobutil-add-output.\ncount_func () \n{ \n    echo \"Value of segmentedbam_file: '${segmentedbam_file}'\";\n    echo \"Chromosome being counted '${chr}'\";\n    dx download \"${segmentedbam_file}\";\n    readcount=$(samtools view -c \"${segmentedbam_file_name}\");\n    printf \"${chr}:\\t%s\\n\" \"${readcount}\" &gt; \"${segmentedbam_file_prefix}.txt\";\n    readcount_file=$(dx upload \"${segmentedbam_file_prefix}\".txt --brief);\n    dx-jobutil-add-output counts_txt \"${readcount_file}\" --class=file\n}\n\n\nsum_reads\nThe main entry point triggers this subjob, providing the output of count_func as an input. This entry point gathers all the files generated by the count_func jobs and sums them.\n\nThis function returns read_sum_file as the entry point output.\nsum_reads () \n{ \n    set -e -x -o pipefail;\n    printf \"Value of read file array %s\" \"${readfiles[@]}\";\n    echo \"Filename: ${filename}\";\n    echo \"Summing values in files and creating output read file\";\n    for read_f in \"${readfiles[@]}\";\n    do\n        echo \"${read_f}\";\n        dx download \"${read_f}\" -o - &gt;&gt; chromosome_result.txt;\n    done;\n    count_file=\"${filename}_chromosome_count.txt\";\n    total=$(awk '{s+=$2} END {print s}' chromosome_result.txt);\n    echo \"Total reads: ${total}\" &gt;&gt; \"${count_file}\";\n    readfile_name=$(dx upload \"${count_file}\" --brief);\n    dx-jobutil-add-output read_sum_file \"${readfile_name}\" --class=file\n}"
					}
					
				
			
		
			
				
					,
					
					"distributed-bash-samtools-count-distr-chr-region-sh": {
						"id": "distributed-bash-samtools-count-distr-chr-region-sh",
						"title": "Distributed by Region (sh)",
						"categories": "distributed, bash",
						"url": " /distributed/bash/samtools_count_distr_chr_region_sh/",
						"content": "Documentation to create a distributed applet can be found on the Developer Tutorials wiki page. This readme will focus on the details of this example.\n\nEntry Points\n\nDistributed bash-interpreter apps use bash functions to declare entry points. Entry points are executed as subjobs on new workers with their own respective system requirements. This app has the following entry points specified as bash functions:\n\n\n  main\n  count_func\n  sum_reads\n\n\nmain\nThe main function takes the initial *.bam, generates an index *.bai if needed, and obtains the list of regions from the *.bam file. Every 10 regions will be sent, as input, to the count_func entry point using dx-jobutil-new-job command.\n  regions=$(samtools view -H \"${mappings_sorted_bam_name}\" | grep \"\\@SQ\" | sed 's/.*SN:\\(\\S*\\)\\s.*/\\1/')\n\n  echo \"Segmenting into regions\"\n  count_jobs=()\n  counter=0\n  temparray=()\n  for r in $(echo $regions); do\n    if [[ \"${counter}\" -ge 10 ]]; then\n      echo \"${temparray[@]}\"\n      count_jobs+=($(dx-jobutil-new-job -ibam_file=\"${mappings_sorted_bam}\" -ibambai_file=\"${mappings_sorted_bai}\" \"${temparray[@]}\" count_func))\n      temparray=()\n      counter=0\n    fi\n    temparray+=(\"-iregions=${r}\") # Here we add to an array of -i&lt;parameter&gt;'s\n    counter=$((counter+1))\n  done\n\n  if [[ counter -gt 0 ]]; then # Previous loop will miss last iteration  if its &lt; 10\n    echo \"${temparray[@]}\"\n    count_jobs+=($(dx-jobutil-new-job -ibam_file=\"${mappings_sorted_bam}\" -ibambai_file=\"${mappings_sorted_bai}\" \"${temparray[@]}\" count_func))\n  fi\n\n\nJob outputs from the count_func entry point are referenced as Job Based Object References (JBOR) and used as inputs for the sum_reads entry point.\n  echo \"Merge count files, jobs:\"\n  echo \"${count_jobs[@]}\"\n  readfiles=()\n  for count_job in \"${count_jobs[@]}\"; do\n    readfiles+=(\"-ireadfiles=${count_job}:counts_txt\")\n  done\n  echo \"file name: ${sorted_bamfile_name}\"\n  echo \"Set file, readfile variables:\"\n  echo \"${readfiles[@]}\"\n  countsfile_job=$(dx-jobutil-new-job -ifilename=\"${mappings_sorted_bam_prefix}\" \"${readfiles[@]}\" sum_reads)\n\n\nJob outputs of the sum_reads entry point is used as the output of the main entry point via JBOR reference in the dx-jobutil-add-output command.\n  echo \"Specifying output file\"\n  dx-jobutil-add-output counts_txt \"${countsfile_job}:read_sum\" --class=jobref\n}\n\n\ncount_func\nThis entry point performs a SAMtools count of the 10 regions passed as input. This execution will be run on a new worker. As a result variables from other functions (e.g. main()) will not be accessible here.\n\nOnce the output file with counts is created, it is uploaded to the platform and assigned as the entry point’s job output counts_txt via the command dx-jobutil-add-output.\ncount_func() {\n\n  set -e -x -o pipefail\n\n  echo \"Value of bam_file: '${bam_file}'\"\n  echo \"Value of bambai_file: '${bambai_file}'\"\n  echo \"Regions being counted '${regions[@]}'\"\n\n\n  dx-download-all-inputs\n\n\n  mkdir workspace\n  cd workspace || exit\n  mv \"${bam_file_path}\" .\n  mv \"${bambai_file_path}\" .\n  outputdir=\"./out/samtool/count\"\n  mkdir -p \"${outputdir}\"\n  samtools view -c \"${bam_file_name}\" \"${regions[@]}\" &gt;&gt; \"${outputdir}/readcounts.txt\"\n\n\n  counts_txt_id=$(dx upload \"${outputdir}/readcounts.txt\" --brief)\n  dx-jobutil-add-output counts_txt \"${counts_txt_id}\" --class=file\n}\n\n\nsum_reads\nThe main entry point triggers this subjob, providing the output of count_func as an input JBOR. This entry point gathers all the readcount.txt files generated by the count_func jobs and sums the totals.\n\nThis entry point returns read_sum as a JBOR, which is then referenced as job output.\nsum_reads() {\n\n  set -e -x -o pipefail\n  echo \"$filename\"\n\n  echo \"Value of read file array '${readfiles[@]}'\"\n  dx-download-all-inputs\n  echo \"Value of read file path array '${readfiles_path[@]}'\"\n\n  echo \"Summing values in files\"\n  readsum=0\n  for read_f in \"${readfiles_path[@]}\"; do\n    temp=$(cat \"$read_f\")\n    readsum=$((readsum + temp))\n  done\n\n  echo \"Total reads: ${readsum}\" &gt; \"${filename}_counts.txt\"\n\n  read_sum_id=$(dx upload \"${filename}_counts.txt\" --brief)\n  dx-jobutil-add-output read_sum \"${read_sum_id}\" --class=file\n\n\nIn the main function, the output is referenced\n  echo \"Specifying output file\"\n  dx-jobutil-add-output counts_txt \"${countsfile_job}:read_sum\" --class=jobref\n}"
					}
					
				
			
		
			
				
					,
					
					"bash-samtools-count-catmkfifo-sh": {
						"id": "bash-samtools-count-catmkfifo-sh",
						"title": "Mkfifo and dx cat",
						"categories": "bash",
						"url": " /bash/samtools_count_catmkfifo_sh/",
						"content": "This applet performs a SAMtools count on an input file while minimizing disk usage. For additional details on using FIFO (named pipes) special files, run the command man fifo in your shell.\n\n Warning: Named pipes require BOTH a stdin and stdout or they will block a process. In these examples, we place incomplete named pipes in background processes so the foreground script process does not block.\n\nTo approach this use case, let’s focus on what we want our applet to do:\n\n  Stream the BAM file from the platform to a worker.\n  As the BAM is streamed, count the number of reads present.\n  Output the result into a file.\n  Stream the result file to the platform.\n\n\nStream BAM file from the platform to a worker\nFirst, we establish a named pipe on the worker. Then, we stream to stdin of the named pipe and download the file as a stream from the platform using dx cat.\n  mkdir workspace\n  mappings_fifo_path=\"workspace/${mappings_bam_name}\"\n  mkfifo \"${mappings_fifo_path}\" # FIFO file is created\n  dx cat \"${mappings_bam}\" &gt; \"${mappings_fifo_path}\" &amp;\n  input_pid=\"$!\"\n\n\n\n  \n    \n      FIFO\n      stdin\n      stdout\n    \n  \n  \n    \n      BAM file\n      YES\n      NO\n    \n  \n\n\nOutput BAM file read count\nNow that we have created our FIFO special file representing the streamed BAM, we can just call the samtools command as we normally would. The samtools command reading the BAM would provide our BAM FIFO file with a stdout. However, keep in mind that we want to stream the output back to the platform. We must create a named pipe representing our output file too.\n  mkdir -p ./out/counts_txt/\n\n  counts_fifo_path=\"./out/counts_txt/${mappings_bam_prefix}_counts.txt\"\n\n  mkfifo \"${counts_fifo_path}\" # FIFO file is created, readcount.txt\n  samtools view -c \"${mappings_fifo_path}\" &gt; \"${counts_fifo_path}\" &amp;\n  process_pid=\"$!\"\n\n\n\n  \n    \n      FIFO\n      stdin\n      stdout\n    \n  \n  \n    \n      BAM file\n      YES\n      YES\n    \n    \n      output file\n      YES\n      NO\n    \n  \n\n\nThe directory structure created here (~/out/counts_txt) is required to use the dx-upload-all-outputs command in the next step.\nAll files found in the path ~/out/&lt;output name&gt; will be uploaded to the corresponding &lt;output name&gt; specified in the dxapp.json.\n\nStream the result file to the platform\nCurrently, we’ve established a stream from the platform, piped the stream into a samtools command, and finally outputting the results to another named pipe. However, our background process is still blocked since we lack a stdout for our output file. Luckily, creating an upload stream to the platform will resolve this.\n\nWe can upload as a stream to the platform using the commands dx-upload-all-outputs or dx upload -. Make sure to specify --buffer-size if needed.\n  mkdir -p ./out/counts_txt/\n\n  counts_fifo_path=\"./out/counts_txt/${mappings_bam_prefix}_counts.txt\"\n\n  mkfifo \"${counts_fifo_path}\" # FIFO file is created, readcount.txt\n  samtools view -c \"${mappings_fifo_path}\" &gt; \"${counts_fifo_path}\" &amp;\n  process_pid=\"$!\"\n\n\n\n  \n    \n      FIFO\n      stdin\n      stdout\n    \n  \n  \n    \n      BAM file\n      YES\n      YES\n    \n    \n      output file\n      YES\n      YES\n    \n  \n\n\n Note: Alternatively, dx upload - can upload directly from stdin. In this example, we would no longer need to have the directory structure required for dx-upload-all-outputs.\n\n Warning: When uploading a file that exists on disk, dx upload is aware of the file size and automatically handles any cloud service provider upload chunk requirements. When uploading as a stream, the file size is not automatically known and dx upload uses default parameters. While these parameters are fine for most use cases, you may need to specify upload part size with the --buffer-size option.\n\nWait for background processes\nNow that our background processes are no longer blocking the rest of the applet’s execution, we simply wait in the foreground for those processes to finish.\n  wait -n  # \"$input_pid\"\n  wait -n  # \"$process_pid\"\n  wait -n  # \"$upload_pid\"\n\n\n Note: If we didn’t wait the app script would running in the foreground would finish and terminate the job! We wouldn’t want that.\n\nHow is the SAMtools dependency provided?\nThe SAMtools compiled binary is placed directly in the &lt;applet dir&gt;/resources directory. Any files found in the resources/ directory will be uploaded so that they will be present in the worker’s root directory. In our case:\n├── Applet dir\n│   ├── src\n│   ├── dxapp.json\n│   ├── resources\n│       ├── usr\n│           ├── bin\n│               ├── &lt; samtools binary &gt;\n\nWhen this applet is run on a worker, the resources/ folder will be placed in the worker’s root directory /:\n\n/\n├── usr\n│   ├── bin\n│       ├── &lt; samtools binary &gt;\n├── home\n│   ├── dnanexus\n\n\n/usr/bin is part of the $PATH variable, so we can reference the samtools command directly in our script as samtools view -c ..."
					}
					
				
			
		
			
				
					,
					
					"bash-samtools-count-binary-sh": {
						"id": "bash-samtools-count-binary-sh",
						"title": "Precompiled Binary",
						"categories": "bash",
						"url": " /bash/samtools_count_binary_sh/",
						"content": "This tutorial showcases packaging a precompiled binary in the resources/ directory of an app(let).\n\nPrecompiling a Binary\nIn this applet, the SAMtools binary was precompiled on an Ubuntu 14.04 machine. A user can do this compilation on an Ubuntu 14.04 machine of their own, or they can utilize the Cloud Workstation app to build and compile a binary. On the Cloud Workstation, the user can download the SAMtools source code and compile it in the worker environment, ensuring that the binary will run on future workers.\n\nSee Cloud Workstation in the App library for more information.\n\nResources Directory\nThe SAMtools precompiled binary is placed in the &lt;applet dir&gt;/resources/ directory. Any files found in the resources/ directory will be packaged, uploaded to the platform, and then unpackaged in the root directory \\ of the worker. In our case, the resources/ dir is structured as follows:\n├── Applet dir\n│   ├── src\n│   ├── dxapp.json\n│   ├── resources\n│       ├── usr\n│           ├── bin\n│               ├── &lt; samtools binary &gt;\n\nWhen this applet is run on a worker, the resources/ directory will be placed in the worker’s root directory /:\n/\n├── usr\n│   ├── bin\n│       ├── &lt; samtools binary &gt;\n├── home\n│   ├── dnanexus\n│   \t├── applet script\n\nWe are able to access the SAMtools command because the respective binary is visible from the default $PATH variable. The directory /usr/bin/ is part of the $PATH variable, so in our script we can reference the samtools command directly:\nsamtools view -c \"${mappings_bam_name}\" &gt; \"${mappings_bam_prefix}.txt\"\n\n\nSee the resources/ directory wiki page for more information."
					}
					
				
			
		
			
				
					,
					
					"bash-samtools-count-bashhelper-sh": {
						"id": "bash-samtools-count-bashhelper-sh",
						"title": "Bash Helpers",
						"categories": "bash",
						"url": " /bash/samtools_count_bashhelper_sh/",
						"content": "This applet performs a basic SAMtools count with the aid of bash helper variables.\n\nDownload BAM Files\nInput files are downloaded using the dx-download-all-inputs command. The dx-download-all-inputs command will go through all inputs and download into folders with the pattern\n/home/dnanexus/in/[VARIABLE]/[file or subfolder with files].\ndx-download-all-inputs\n\n\nCreate Output Directory\n\nWe create an output directory in preparation for dx-upload-all-outputs DNAnexus command in the (Upload Results)[#Upload Result] section.\nmkdir -p out/counts_txt\n\n\nRun SAMtools View\n\nAfter executing the dx-download-all-inputs command, there are three helper variables created to aid in scripting. For this applet, the input variable name mappings_bam with platform filename my_mappings.bam will have a helper variables:\n# [VARIABLE]_path the absolute string path to the file.\n$ echo $mappings_bam_path\n/home/dnanexus/in/mappings_bam/my_mappings.bam\n# [VARIABLE]_prefix the file name minus the longest matching pattern in the dxapp.json file\n$ echo $mappings_bam_prefix\nmy_mappings\n# [VARIABLE]_name the file name from the platform\n$ echo $mappings_bam_name\nmy_mappings.bam\n\nWe use the bash helper variable mappings_bam_path to reference the location of a file after it has been downloaded using dx-download-all-inputs.\nsamtools view -c \"${mappings_bam_path}\" &gt; out/counts_txt/\"${mappings_bam_prefix}.txt\"\n\n\nUpload Result\n\nWe use the dx-upload-all-outputs command to upload data to the platform and specify\nit as the job’s output. The dx-upload-all-outputs command expects to find file paths matching the pattern\n/home/dnanexus/out/[VARIABLE]/*. It will upload matching files and then associate them as the output corresponding to [VARIABLE]. In this case,\nthe output is called counts_txt. Above we created the folders, and we can now place the outputs there.  \ndx-upload-all-outputs"
					}
					
				
			
		
			
				
					,
					
					"bash-samtools-count": {
						"id": "bash-samtools-count",
						"title": "SAMtools count",
						"categories": "bash",
						"url": " /bash/samtools_count/",
						"content": "This applet performs a basic samtools view -c {bam} command, referred to as “SAMtools count”, on the DNAnexus platform.\n\nDownload BAM Files\nFor bash scripts, inputs to a job execution become environment variables. The inputs from our dxapp.json file are formatted as shown below:\n{\n  \"inputSpec\": [\n    {\n      \"name\": \"mappings_bam\",\n      \"label\": \"Mapping\",\n      \"class\": \"file\",\n      \"patterns\": [\"*.bam\"],\n      \"help\": \"BAM format file.\"\n    }\n  ]\n}\n\nThe object mappings_bam, a DNAnexus link\ncontaining the file ID of that file, will be available as an environmental variable in the applet’s execution. Use the command dx download to download the BAM file. By default, when we download a file,\nwe will keep the filename of the object on the platform.\ndx download \"${mappings_bam}\"\n\n\nSAMtools Count\nHere, we use the bash helper variable mappings_bam_name. For file inputs,\nthe DNAnexus platform creates a bash variable [VARIABLE]_name that holds a string representing\nthe filename of the object on the platform; because we downloaded the file using default parameters, this will be the filename of the object on this\nworker as well. We use another helper variable, [VARIABLE]_prefix, the filename\nof the object minus any suffixes specified in the input field patterns. From the input spec above,\nthe only pattern present is '[\"*.bam\"]', so the platform will remove the trailing “.bam” and create the helper variable [VARIABLE]_prefix for our use.\nreadcount=$(samtools view -c \"${mappings_bam_name}\")\necho \"Total reads: ${readcount}\" &gt; \"${mappings_bam_prefix}.txt\"\n\n\nUpload Result\nUse dx upload command to upload data to the platform. This will upload the file into the\njob container, a temporary project that holds onto files associated\nwith the job. When running the command dx upload with the flag --brief, the command will return just the\nfile ID.\ncounts_txt_id=$(dx upload \"${mappings_bam_prefix}.txt\" --brief)\n\n\n Note: While job containers are an integral part of the execution process, a deeper discussion goes out of scope of a basic tutorial. Review the Containers for Execution wiki page for more information.\n\nAssociate With Output\nThe output of an applet must be declared before the applet is even built. Looking back to the dxapp.json file, we see the following:\n{\n  \"name\": \"counts_txt\",\n  \"class\": \"file\",\n  \"label\": \"Read count file\",\n  \"patterns\": [\n    \"*.txt\"\n  ],\n  \"help\": \"Output file with Total reads as the first line.\"\n}\n\nWe declared a file type output named counts_txt. In the applet script, we must tell the system what file should be associated with the output counts_txt. On job completion, usually at the end of the script, this file will be copied from the temporary job container to the project that launched the job.\ndx-jobutil-add-output counts_txt \"${counts_txt_id}\" --class=file"
					}
					
				
			
		
			
				
					,
					
					"web-app-bash-r-shiny-web-app": {
						"id": "web-app-bash-r-shiny-web-app",
						"title": "R Shiny Example Web App",
						"categories": "web_app, bash",
						"url": " /web_app/bash/r-shiny-web-app/",
						"content": "This is an example web applet that demonstrates how to build and run an R Shiny application on DNAnexus.\n\nCreating the web application\nInside the dxapp.json, you would add \"httpsApp\": {\"ports\":[443], \"shared_access\": \"VIEW\"} to tell the worker to expose this port.\n\nR Shiny needs two scripts, server.R and ui.R, which should be under resources/home/dnanexus/my_app/. When a job starts based on this applet, the resources directory is copied onto the worker, and since the ~/ path on the worker is /home/dnanexus, that means you now have ~/my_app with those two scripts inside.\n\nFrom the main applet script code.sh, we simply start shiny pointing to ~/my_app repo, serving its mini-application on port 443.\n\nmain() {\n  R -e \"shiny::runApp('~/my_app', host='0.0.0.0', port=443)\"\n}\n\n\nNote that for all web apps, if everything is running smoothly and no errors are encountered (the ideal case), the line of code that starts the server will keep it running forever. The applet stops only when it is terminated. This also means that any lines of code after the server starts will not be executed.\n\nModifying this example for your own applet\nTo make your own applet with R Shiny, simply copy the source code from this example and modify server.R and ui.R inside resources/home/dnanexus/my_app.\n\nHow to rebuild the shiny asset\ndx build_asset shiny-asset\n\nThis will output a record ID record-xxxx that you can then put into the applet’s dxapp.json in place of the existing one:\n\"runSpec\": {\n\t...\n\t\"assetDepends\": [\n    {\n      \"id\": \"record-xxxx\n    }\n  ]\n\t...\n}\n\n\nBuild the applet\nNow build and run the applet itself:\ndx build -f dash-web-app\ndx run dash-web-app\n\nOnce it spins up, you can go to that job’s designated URL based on its job ID, https://job-xxxx.dnanexus.cloud/, to see the result."
					}
					
				
			
		
			
				
					,
					
					"python-pysam-count": {
						"id": "python-pysam-count",
						"title": "Pysam",
						"categories": "python",
						"url": " /python/pysam_count/",
						"content": "This applet performs a SAMtools count on an input BAM using Pysam, a python wrapper for SAMtools.\n\nHow is Pysam provided?\n\nPysam is provided through a pip install using the pip package manager in the dxapp.json’s runSpec.execDepends property:\n\n\n{\n \"runSpec\": {\n    ...\n    \"execDepends\": [\n      {\"name\": \"pysam\",\n         \"package_manager\": \"pip\",\n         \"version\": \"0.9.1.4\"\n      }\n    ]\n    ...\n }\n\n\nThe execDepends value is a JSON array of dependencies to resolve before the applet source code is run. In this applet, we specify pip as our package manager and pysam version 0.9.1.4 as the dependency to resolve. Pysam is installed to /usr/local/lib/python2.7/dist-packages and can be imported by our python script.\n\nDownloading Inputs\nThe fields mappings_sorted_bam and mappings_sorted_bai are passed to the main function as parameters for our job. These parameters are dictionary objects with key-value pair {\"$dnanexus_link\": \"&lt;file&gt;-&lt;xxxx&gt;\"}. We handle file objects from the platform through DXFile handles. If an index file is not supplied, then a *.bai index will be created.\n    print mappings_sorted_bai\n    print mappings_sorted_bam\n\n    mappings_sorted_bam = dxpy.DXFile(mappings_sorted_bam)\n    sorted_bam_name = mappings_sorted_bam.name\n    dxpy.download_dxfile(mappings_sorted_bam.get_id(),\n                         sorted_bam_name)\n    ascii_bam_name = unicodedata.normalize(  # Pysam requires ASCII not Unicode string.\n        'NFKD', sorted_bam_name).encode('ascii', 'ignore')\n\n    if mappings_sorted_bai is not None:\n        mappings_sorted_bai = dxpy.DXFile(mappings_sorted_bai)\n        dxpy.download_dxfile(mappings_sorted_bai.get_id(),\n                             mappings_sorted_bai.name)\n    else:\n        pysam.index(ascii_bam_name)\n\n\nWorking with Pysam\nPysam provides several methods that mimic SAMtools commands. In our applet example, we want to focus only on canonical chromosomes. Pysam’s object representation of a BAM file is pysam.AlignmentFile.\n    mappings_obj = pysam.AlignmentFile(ascii_bam_name, \"rb\")\n    regions = get_chr(mappings_obj, canonical_chr)\n\n\nThe helper function get_chr\ndef get_chr(bam_alignment, canonical=False):\n    \"\"\"Helper function to return canonical chromosomes from SAM/BAM header\n\n    Arguments:\n        bam_alignment (pysam.AlignmentFile): SAM/BAM pysam object\n        canonical (boolean): Return only canonical chromosomes\n    Returns:\n        regions (list[str]): Region strings\n    \"\"\"\n    regions = []\n    headers = bam_alignment.header\n    seq_dict = headers['SQ']\n\n    if canonical:\n        re_canonical_chr = re.compile(r'^chr[0-9XYM]+$|^[0-9XYM]')\n        for seq_elem in seq_dict:\n            if re_canonical_chr.match(seq_elem['SN']):\n                regions.append(seq_elem['SN'])\n    else:\n        regions = [''] * len(seq_dict)\n        for i, seq_elem in enumerate(seq_dict):\n            regions[i] = seq_elem['SN']\n\n    return regions\n\n\nOnce we establish a list of canonical chromosomes, we can iterate over them and perform Pysam’s version of samtools view -c, pysam.AlignmentFile.count.\n    total_count = 0\n    count_filename = \"{bam_prefix}_counts.txt\".format(\n        bam_prefix=ascii_bam_name[:-4])\n\n    with open(count_filename, \"w\") as f:\n        for region in regions:\n            temp_count = mappings_obj.count(region=region)\n            f.write(\"{region_name}: {counts}\\n\".format(\n                region_name=region, counts=temp_count))\n            total_count += temp_count\n\n        f.write(\"Total reads: {sum_counts}\".format(sum_counts=total_count))\n\n\nUploading Outputs\nOur summarized counts are returned as the job output. We use the dx-toolkit python SDK’s dxpy.upload_local_file function to upload and generate a DXFile corresponding to our tabulated result file.\n    counts_txt = dxpy.upload_local_file(count_filename)\n    output = {}\n    output[\"counts_txt\"] = dxpy.dxlink(counts_txt)\n\n    return output\n\n\nPython job outputs have to be a dictionary of key-value pairs, with the keys being job output names as defined in the dxapp.json file and the values being the output values for corresponding output classes. For files, the output type is a DXLink. We use the dxpy.dxlink function to generate the appropriate DXLink value."
					}
					
				
			
		
			
				
					,
					
					"python-web-app-dash-web-app": {
						"id": "python-web-app-dash-web-app",
						"title": "Dash Example Web App",
						"categories": "python, web_app",
						"url": " /python/web_app/dash-web-app/",
						"content": "This is an example web app made with Dash, which in turn uses Flask underneath.\n\nCreating the web application\nAfter configuring an app with Dash, we start the server on port 443.\napp.run_server(host='0.0.0.0', port=443)\n\nInside the dxapp.json, you would add \"httpsApp\": {\"ports\":[443], \"shared_access\": \"VIEW\"} to tell the worker to expose this port.\n\nNote that for all web apps, if everything is running smoothly and no errors are encountered (the ideal case), the line of code that starts the server will keep it running forever. The applet stops only when it is terminated. This also means that any lines of code after the server starts will not be executed.\n\nThe rest of these instructions apply to building any applet with dependencies stored in an asset.\n\nCreating an applet on DNAnexus\nSource dx-toolkit and log in, then run dx-app-wizard with default options.\n\nCreating the asset\ndash-asset specifies all the packages and versions we need.\nWe take these from the Dash installation guide (https://dash.plot.ly/installation)\npip install dash==0.39.0  # The core dash backend\npip install dash-html-components==0.14.0  # HTML components\npip install dash-core-components==0.44.0  # Supercharged components\npip install dash-table==3.6.0  # Interactive DataTable component (new!)\npip install dash-daq==0.1.0  # DAQ components (newly open-sourced!)\n\n\nWe put these into dash-asset/dxasset.json:\n{\n  ...\n  \"execDepends\": [\n    {\"name\": \"dash\", \"version\":\"0.39.0\", \"package_manager\": \"pip\"},\n\t\t{\"name\": \"dash-html-components\", \"version\":\"0.14.0\", \"package_manager\": \"pip\"},\n\t\t{\"name\": \"dash-core-components\", \"version\":\"0.44.0\", \"package_manager\": \"pip\"},\n\t\t{\"name\": \"dash-table\", \"version\":\"3.6.0\", \"package_manager\": \"pip\"},\n\t\t{\"name\": \"dash-daq\", \"version\":\"0.1.0\", \"package_manager\": \"pip\"}\n  ],\n\t...\n}\n\n\nBuild the asset:\ndx build_asset dash-asset\n\n\nUse the asset from the applet\nAdd this asset to the applet’s dxapp.json:\n\"runSpec\": {\n\t...\n\t\"assetDepends\": [\n    {\n      \"id\": \"record-xxxx\n    }\n  ]\n\t...\n}\n\n\nBuild the applet\nNow build and run the applet itself:\ndx build -f dash-web-app\ndx run dash-web-app\n\nYou can always use dx ssh job-xxxx to ssh into the worker and inspect what’s going on or experiment with quick changes \nThen go to that job’s special URL https://job-xxxx.dnanexus.cloud/ and see the result!\n\nOptional local testing\nThe main code is in dash-web-app/resources/home/dnanexus/my_app.py with a local launcher script called local_test.py in the same folder. This allows us to launch the same core code in the applet locally to quickly iterate. This is optional because you can also do all testing on the platform itself.\n\nInstall locally the same libraries listed above.\n\nTo launch the web app locally:\ncd dash-web-app/resources/home/dnanexus/\npython local_test.py\n\nOnce it spins up, you can go to that job’s designated URL based on its job ID, https://job-xxxx.dnanexus.cloud/, to see the result."
					}
					
				
			
		
			
				
					,
					
					"next-steps": {
						"id": "next-steps",
						"title": "Next Steps",
						"categories": "",
						"url": " /next-steps/",
						"content": "Almost There!\n\nNow that you’re almost at the end of the Getting Started series, you’re ready to start developing, building and publishing apps on the DNAnexus Platform! As you continue your journey remember; there’s always more to learn and new techniques to explore when developing applications, so come back to this page as needed.\n\nSo what’s Next? Depending what you’re looking for, there are a couple different routes you can go from here.\n\nMaking app(let)s\n\nIf you’re looking to start cranking out app(let)s as soon as possible, I recommend taking a look the Tutorials page. Our tutorial will have a collection of best practices and how-to’s when it comes to app development on the DNAnexus platform. You can take concepts from our simple examples and apply them to your specific use-case.\n\nBecome a power user\n\nIf you’re looking to dive into more advanced app(let) techniques we highly recommend taking a look at the resources page. You’ll find several useful links. Two I’d like to point here:\n\nDNAnexus Wiki\n\nOur wiki is our platforms living documentation. For more advanced users, it is home to our API specification. Our platform SDK and even our platform web-page both construct and send API request to endpoints found on this page. Understanding what API resources are available to you is an essential step to becoming a power user on the DNAnexus platform.\n\nDNAnexus Autodocs\n\nThis contains our auto-generated documentation (with source code) of our platform SDK, dx-toolkit. A great starting point for learning how to programmatically interact with our API is to see how we interact with our API. dx-toolkit is completely open-source and available on GitHub for your perusal. A lot of the patterns you find in our SDK can be helpful in your production code."
					}
					
				
			
		
			
				
					,
					
					"overview": {
						"id": "overview",
						"title": "DNAnexus Tutorials",
						"categories": "",
						"url": " /overview/",
						"content": "Prerequisites\n\nWe assume you already have a DNAnexus account and are looking to create app(let)s on the platform. If you need to create an account please visit our sign up page and consider giving the website quickstart a read. Additionally, the DNAnexus SDK, dx-toolkit, must be installed to run several of the commands used in the tutorials and examples.\n\nInstall the DNAnexus SDK\n\nThe DNAnexus SDK, dx-toolkit, provides a collection of tools to aid in the app(let) development process. While it is possible to develop and publish an app(let) using our API directly, we recommend using our SDK:\n\nDownload dx-toolkit\n\ndx-toolkit can be found on our wiki Downloads page. Alternatively, it can be built from the github source code.\n\n~ $ curl -O https://wiki.dnanexus.com/images/files/dx-toolkit-v0.228.0-osx.tar.gz\n~ $ tar -xvf dx-toolkit-v0.228.0-osx.tar.gz\n  # A folder dx-toolkit/ will be created in the current directory\n\n\nsource the dx-toolkit environment.\n\n~ $ source dx-toolkit/environment\n  # You are now ready to use dx commands\n~ $ dx --version\n  # Downloaded dx-toolkit version outputs\n\n\nLogin to the platform\n\nFrom the CLI use dx login. You may wish to login with a token.\n\n~ $ dx login\n  Username:  # Platform username\n  Password:  # Platform password\n  # If this is your first-time login from the Command Line Interface then you will be prompted to select a project\n  Note: Use dx select --level VIEW or dx select --public to select from projects for\n  which you only have VIEW permissions.\n\n  Available projects (CONTRIBUTE or higher):\n  0) DNAnexus Example (CONTRIBUTE)\n  1) My Research Project (ADMINISTER)\n  \n  Pick a numbered choice or \"m\" for more options: 1\n  Setting current project to: My Research Project\n\n\nCreate a test applet\n\ndx-app-wizard generates the directory structure required to build an app. For now, there will be no inputs or outputs we’ll just use the default options when prompted.\n\n~ $ dx-app-wizard\nDNAnexus App Wizard, API v1.0.0\n\nBasic Metadata\n\nPlease enter basic metadata fields that will be used to describe your app.  Optional fields are denoted by options with square brackets.  At the end of this wizard, the files necessary for building your app will be generated from the answers you provide.\n\n  # We'll name our applet \"My First Applet\"\n\n  # When the dx-app-wizard is done it'll construct a full, valid, applet directory.\n  # In this case, we'll have the following directory structure:\n  my_first_applet\n  ├── src\n  │   └── my_first_applet.sh\n  │\n  ├── resources/\n  ├── test/\n  ├── dxapp.json\n  ├── Readme.md\n  └── Readme.developer.md\n# As you read out wiki and go through these tutorials you'll understand the role each file and directory plays in a developed app(let)\n\n\nBuild the sample app(let)\n\n~ $ dx build my_first_applet\n{\"id\": \"applet-xxxx\"} # This is the entity-ID of the applet created in the project\n\n\nIf you are able to perform the steps above, you’re ready to start creating app(let)s! Next time you or a colleague browse your project, they’ll see this newly created applet."
					}
					
				
			
		
			
				
					,
					
					"do-you-need-app": {
						"id": "do-you-need-app",
						"title": "Assessing Your Use Case",
						"categories": "",
						"url": " /do-you-need-app/",
						"content": "Before embarking on a new journey it’s important to ask, “Where am I going?” and “What am I looking for?”. This page takes a step back and assesses alternative solutions and approaches that may get you to the real answer you want.\n\nLook before you leap.\n\nAlternatives to Applets\nBefore you start implementing your own app(let), you should explore the DNAnexus App Library. Our App Library, which we continuously update, contains common bioinformatics tools implemented on the platform. The Apps are robust and generalized in their implementation and often expose many options to the user. There may be cases where you want to make a slight tweak to an existing App, for this reason, we open source our Apps. Below is an example of getting the source code for the BWA-MEM FASTQ Read Mapper using dx find apps, dx describe, and dx get:\n\n~ $ dx find apps | grep \"BWA-MEM FASTQ Read Mapper\"\nx BWA-MEM FASTQ Read Mapper (bwa_mem_fastq_read_mapper), v1.5.4\n  # We will use the unique App name \"bwa_mem_fastq_read_mapper\" to get the app-id\n~ $ dx describe bwa_mem_fastq_read_mapper\nResult 1:\nID                  app-xxxx  # get this ID\nClass               app\nBilled to           org-dnanexus\nName                bwa_mem_fastq_read_mapper\n...\n~ $ dx get app-xxxx\nCreating \"./bwa_mem_fastq_read_mapper\" output directory\nDownloading application data\nUnpacking resources\n# Now we have the populated app(let) directory structure for BWA-MEM\n# In the folder bwa_mem_fastq_read_mapper/\n# We make necessary minor changes and rebuild\ndx build bwa_mem_fastq_read_mapper\n# app-xxxx\n\n\nSwiss Army Knife\n\nFor running quick one-liners on a worker we recommend the swiss army knife.\n\nFor example, you might want to merge overlapping entries in a BED file\nusing a bedtools command. You could create an applet that has a BED file input\nand a merged BED file output, or you could just type the command and run it using\nthe Swiss Army Knife app.\n\nThe Swiss Army Knife app takes an arbitrary list of files and a command string as\ninputs. In commands, it’s often useful to refer to input files using wildcards, like\n*.bed, so the command can be independent of the specific input files. It downloads all the files, runs the command, and uploads any new files. So, we could just add our BED file as an input and type in the appropriate command. From command line using dx run:\n\n~ $ dx run swiss-army-knife\nEntering interactive mode for input selection.\n\nInput:   Command line (cmd)\nClass:   string\n\nEnter string value ('?' for more options)\ncmd: bedtools merge -i *.bed &gt; merged.bed\n\nSelect an optional parameter to set by its # (^D or &lt;ENTER&gt; to finish):\n\n [0] Input files (in)\n [1] Optional Docker image identifier (image)\n\nOptional param #: 0\n\nInput:   Input files (in)\nClass:   array:file\n\nEnter file values, one at a time (^D or &lt;ENTER&gt; to finish, &lt;TAB&gt; twice for compatible files in\ncurrent directory, '?' for more options)\nin[0]: /Bed-Files/target.bed\n# Add more files, confirm inputs, then run.\n\n\nThe Swiss Army Knife app has a number of tools built-in that you can refer to in commands. The list of tools and more information are available on the Swiss Army Knife apps’ page.\n\nCloud Workstation\nFor exploratory development or analysis, we recommend Cloud Workstation. The Cloud Workstation app creates a DNAnexus worker for a specified period of time that\n, if you have configured and enabled SSH, you can SSH into. Useful when you want to:\n\n  Work interactively on a remote computer that has fast\naccess to files stored on the DNAnexus platform\n  Work in an environment that has fast access to the internet\n  Work on a machine with high processing power or high storage.\n\n\nRunning the Cloud Workstation app from the terminal:\n\nSet up SSH access with dx ssh_config:\n\n$ dx ssh_config\nSelect an SSH key pair to use when connecting to DNAnexus jobs. The public key will be saved to your\nDNAnexus account (readable only by you). The private key will remain on this computer.                              \ndx is already configured to use the SSH key pair at:\n    /home/me/.dnanexus_config/ssh_id\n    /home/me/.dnanexus_config/ssh_id.pub\n0) Use this SSH key pair\n1) Select or create another SSH key pair...                                                                                             \nPick a numbered choice: 0\nUpdated public key for user user-xxxx\nYour account has been configured for use with SSH. Use dx run with the --allow-ssh, --ssh,\nor --debug-on options to launch jobs and connect to them.\n\n\nRun the app:\n\n$ dx run app-cloud_workstation --allow-ssh                                                \nSelect an optional parameter to set by its # (^D or &lt;ENTER&gt; to finish):        \n[0] Maximum Session Length (suffixes allowed: s, m, h, d, w, M, y) (max_session_length) [default=\"1h\"]                                 \n[1] Files (fids)                                         \n              \nOptional param #:\nUsing input JSON:\n{}                                                                                                                                      \nConfirm running the executable with this input [Y/n]: y                                                                                 \nCalling app-F40jZqQ9fZF4ggVZ1qGYQz5z with output destination project-xxxx:/                                         \nJob ID: job-xxxx                                                                                                    \nWatch launched job now? [Y/n] n\n\n\nSSH into the job running the app\n\n$ dx ssh job-xxxx\n\n\nThe Cloud Workstation app takes a rather useful approach for exploratory development on the platform. Feel free to look at the app code for Cloud Workstation v1.0.3, which you can obtain with:\ndx get app-F40jZqQ9fZF4ggVZ1qGYQz5z"
					}
					
				
			
		
			
				
					,
					
					"applet-structure": {
						"id": "applet-structure",
						"title": "Basic Applet Structure",
						"categories": "",
						"url": " /applet-structure/",
						"content": "Before you are able to dx build an app(let) you must construct a valid app(let) directory. An app(let) directory skeleton looks like:\nApplet directory\n├── src\n│   └── code.sh  # The applet script\n│\n├── resources/  # (optional directory)\n├── test/\n├── dxapp.json\n├── Readme.md\n└── Readme.developer.md\n\n\nDxapp.json\nThe dxapp.json file defines the interface of the applet and tells the platform how the app(let) should be run. As you may have guessed from the extension *.json an applet is defined in the JSON format. It will contain all the information necessary to build the app(let) on the platform. An example dxapp.json used in our SAMtools count tutorials:\n\n{\n  \"name\": \"samtools_count\",\n  \"title\": \"SAMtools count\",\n  \"summary\": \"Basic SAMtools count implementation\",\n  \"dxapi\": \"1.0.0\",\n  \"version\": \"0.0.1\",\n  \"inputSpec\": [\n    {\n      \"name\": \"mappings_bam\",\n      \"label\": \"Mapping\",\n      \"class\": \"file\",\n      \"patterns\": [\"*.bam\"],\n      \"help\": \"BAM format file.\"\n    }\n  ],\n  \"outputSpec\": [\n    {\n      \"name\": \"counts_txt\",\n      \"class\": \"file\",\n      \"label\": \"Read count file\",\n      \"patterns\": [\n        \"*.txt\"\n      ],\n      \"help\": \"Output file with Total reads as the first line.\"\n    }\n  ],\n...\n\n\nApplet Script\n\nEach applet has a script, written in bash or python2.7, that executes on a worker when a platform job is created. Often, this script just calls other scripts and compiled binaries. These scripts will play a fundamental role as we discuss parallel and distributed applications.\n\nNote: App(let) scripts are executed from the $HOME directory of a worker. Keep this in mind when performing file operations and relative directory changes.\n\nResources Directory\n\nThis directory is optional.\n\nThe contents of this directory are packed and included in an applet when it is built. When an applet is executed the contents are unpacked and placed in the workers \\ root directory. Since an app(let)s’ script is executed from the HOME directory of a worker, it is common to structure the resources directory: resource/home/dnanexus/*, where * represent the contents to unpack.\n\nTest Folder\n\nThis folder is not packaged, nor does it have an impact on app(let) creation. It does, however, touch on the very important topics of Continuous Integration, App development, and App maintenance. These are advanced development topics outside the scope of basic DNAnexus tutorials and examples. I mention this folder only because the dx-app-wizard generates a template to help users write platform compatible Python functional test. Again, these topics are outside the scope of a basic tutorial or app(let) example.\n\nReadme Files\n\nThese are the markdown files that will be rendered and displayed to users of your app(let). In the Readme.md, it is common to include basic usage information such as inputs details, output details, or extended descriptions of parameters. In the Readme.developer.md, it is common to include advanced usage information such as instance type selection, entry point descriptions, or known bugs. You can find more information about Readmes on the wiki App Build Process page."
					}
					
				
			
		
			
				
					,
					
					"app-dev-spec": {
						"id": "app-dev-spec",
						"title": "Development Basics",
						"categories": "",
						"url": " /app-dev-spec/",
						"content": "This page covers essential principles for developing applets. While we won’t cover everything there is, we will provide a good jump-off point. To see these principles in action review the specific tutorials that cover each concept.\n\nIf you’d like a more in-depth tutorial, see the Advanced App Tutorials page in the wiki.\n\nApplets vs Apps\n\nThroughout these tutorials we’ll refer to Apps and Applets interchangeably as App(let)s. During the development process the distinction isn’t as relevant, here is an explanation of the App/Applet distinction\n\n\n  Applets and apps are both executables that can be run in the platform. The main difference is that applets are data objects, which live in projects, while apps do not live in a project and can be published to allow other users to run them.The procedure for creating applets or apps is essentially the same, with the exception of the final build step where you specify whether it should be created as an applet or an app. (You can also convert an existing applet to an app.) As a result, we may use the two terms interchangeably in tutorials like this one.\n\n\n\nInputs and Outputs\n\nYou can envision an app(let) as a black box that takes an input and gives you an output. Throughout our documentation, you may see the term I/O, this just means “Input/Output”. We describe the input and output specification in the dxapp.json file’s inputSpec and outputSpec respectively:\n{\n  \"inputSpec\": [\n    {\n      \"name\": \"mappings_bam\",\n      \"label\": \"Mapping\",\n      \"class\": \"file\",\n      \"patterns\": [\"*.bam\"],\n      \"help\": \"BAM format file.\"\n    }\n  ],\n  \"outputSpec\": [\n    {\n      \"name\": \"counts_txt\",\n      \"class\": \"file\",\n      \"label\": \"Read count file\",\n      \"patterns\": [\n        \"*.txt\"\n      ],\n      \"help\": \"Output file with total counts as the first line.\"\n    }\n  ]\n}\n\nWe offer a great deal of customization with I/O fields. The full details of what properties and values are available in I/O specifications can be found on our Job Input and Output wiki page. When viewing tutorials and examples, build the app(let) for yourself and see how the I/O specification is presented on the platform.\n\nDependencies\n\nBefore your app(let) executes there will certain preconditions you must resolve. One common dependency you’ll see in our tutorials is SAMtools. To resolve this dependency we often use the execDepends property in the dxapp.json.\n{\n    \"execDepends\": [\n      {\"name\": \"samtools\"}\n    ]\n}\n\nAs you read through each tutorial pay close attention to how we resolve the SAMtools dependency. There are various approaches you can take; each better suited for specific use cases.\n\nEntry Points\n\nEntry points are how we distribute work across many workers.\n\nYou can envision entry points as standalone functions that are executed on their own worker, these sub-executions are what we call sub-jobs. They are defined in the dxapp.json file and referenced in your applet script:\n\n  dxapp.json we can set specs for each entry point\n    {\n\"runSpec\": {\n  \"systemRequirements\": {\n    \"main\": {\n      \"instanceType\": \"mem1_ssd1_x4\"\n    },\n    \"count_func\": {\n      \"instanceType\": \"mem1_ssd1_x2\"\n    },\n    \"sum_reads\": {\n      \"instanceType\": \"mem1_ssd1_x4\"\n    }\n  }\n}\n}\n    \n  \n  Python scripts refer to them as decorators on functions\n    @dxpy.entry_point('main')\ndef main(**job_inputs):\n  # do work\n    \n  \n  bash scripts refer to them as the function itself\n    main() {\n  # do work\n}\n    \n    Distributed computing, the primary use case of entry points, takes advantage of the logical separation of work in app(let) scripts. Because each entry point is executed on its own worker you are able to alter instance type and dependencies in the systemRequirements field of your dxapp.json. Our distributed tutorials will showcase how to take advantage of the options entry points offer.\n  \n\n\nA detailed explanation of entry points can be found on the Applets and Entry Points wiki page."
					}
					
				
			
		
	};
</script>
<script src="https://dnanexus.github.io/dnanexus-example-applets/js/lunr.min.js"></script>
<script src="https://dnanexus.github.io/dnanexus-example-applets/js/search.js"></script>
			</div>
		</section>

		<footer>
	<div class="wrapper">
		<ul class="footer-links">
			
				<li><a target="_blank" href="https://twitter.com/dnanexus" class="Twitter-icon">
					
						
		<svg class="twitter" fill="#000000" height="24" viewBox="0 0 24 24" width="24" xmlns="http://www.w3.org/2000/svg"><path d="M22.46,6C21.69,6.35 20.86,6.58 20,6.69C20.88,6.16 21.56,5.32 21.88,4.31C21.05,4.81 20.13,5.16 19.16,5.36C18.37,4.5 17.26,4 16,4C13.65,4 11.73,5.92 11.73,8.29C11.73,8.63 11.77,8.96 11.84,9.27C8.28,9.09 5.11,7.38 3,4.79C2.63,5.42 2.42,6.16 2.42,6.94C2.42,8.43 3.17,9.75 4.33,10.5C3.62,10.5 2.96,10.3 2.38,10C2.38,10 2.38,10 2.38,10.03C2.38,12.11 3.86,13.85 5.82,14.24C5.46,14.34 5.08,14.39 4.69,14.39C4.42,14.39 4.15,14.36 3.89,14.31C4.43,16 6,17.26 7.89,17.29C6.43,18.45 4.58,19.13 2.56,19.13C2.22,19.13 1.88,19.11 1.54,19.07C3.44,20.29 5.7,21 8.12,21C16,21 20.33,14.46 20.33,8.79C20.33,8.6 20.33,8.42 20.32,8.23C21.16,7.63 21.88,6.87 22.46,6Z" /></svg>
	

					
					</a></li>
			
				<li><a target="_blank" href="https://www.youtube.com/channel/UCoQHyIrWjXP675NR-QNzEnQ" class="YouTube-icon">
					
						
		<svg class="youtube" fill="#000000" height="24" viewBox="0 0 24 24" width="24" xmlns="http://www.w3.org/2000/svg"><path d="M10,16.5V7.5L16,12M20,4.4C19.4,4.2 15.7,4 12,4C8.3,4 4.6,4.19 4,4.38C2.44,4.9 2,8.4 2,12C2,15.59 2.44,19.1 4,19.61C4.6,19.81 8.3,20 12,20C15.7,20 19.4,19.81 20,19.61C21.56,19.1 22,15.59 22,12C22,8.4 21.56,4.91 20,4.4Z" /></svg>
	

					
					</a></li>
			
				<li><a target="_blank" href="https://www.dnanexus.com/" class="DNAnexus-icon">
					
						
		<svg class="dnanexus-social-logo-svg" fill="#000000" height="24" viewBox="0 0 24 24" width="24" xmlns="http://www.w3.org/2000/svg"><g xmlns="http://www.w3.org/2000/svg" id="svgg" transform="scale(0.050) translate(75)"><path id="path0" d="M279.476 94.035 C 278.756 95.911,279.533 96.131,280.637 94.364 C 281.255 93.374,281.308 92.800,280.782 92.800 C 280.324 92.800,279.737 93.356,279.476 94.035 M130.623 126.122 C 129.855 126.434,129.855 126.743,130.623 127.579 C 131.691 128.738,133.944 127.603,133.144 126.309 C 132.671 125.544,132.146 125.505,130.623 126.122 M75.608 179.187 C 75.018 180.142,75.767 181.600,76.847 181.600 C 77.821 181.600,77.827 179.435,76.855 178.834 C 76.446 178.581,75.884 178.740,75.608 179.187 M123.473 221.460 C 122.809 222.523,122.847 223.234,123.622 224.295 C 124.599 225.630,124.687 225.618,125.177 224.071 C 125.958 221.613,124.639 219.593,123.473 221.460 M302.143 266.968 C 301.902 267.358,301.997 268.148,302.353 268.724 C 302.845 269.519,303.142 269.541,303.590 268.816 C 303.915 268.290,303.821 267.501,303.381 267.061 C 302.933 266.613,302.388 266.572,302.143 266.968 M73.776 269.152 C 72.773 270.210,72.258 271.265,72.633 271.497 C 73.270 271.890,74.926 270.385,76.298 268.165 C 77.328 266.498,75.681 267.142,73.776 269.152 M202.868 279.710 C 202.236 281.357,203.662 283.065,204.746 281.960 C 205.209 281.489,205.093 280.659,204.441 279.776 C 203.462 278.452,203.353 278.448,202.868 279.710 M50.856 296.709 C 50.615 297.099,50.912 297.829,51.515 298.330 C 52.353 299.025,52.717 298.970,53.051 298.099 C 53.735 296.317,51.840 295.117,50.856 296.709 M178.767 300.053 C 178.477 300.522,178.432 301.099,178.667 301.333 C 179.240 301.906,180.800 300.897,180.800 299.953 C 180.800 298.930,179.419 298.999,178.767 300.053 M176.517 301.185 C 175.839 302.950,176.696 304.424,177.773 303.347 C 178.282 302.838,178.293 302.066,177.802 301.190 C 177.084 299.906,177.007 299.906,176.517 301.185 M242.583 307.350 C 242.757 307.873,243.328 308.443,243.851 308.617 C 244.435 308.812,244.614 308.446,244.315 307.667 C 243.757 306.213,242.103 305.910,242.583 307.350 M154.278 323.113 C 152.745 324.605,152.026 327.164,153.200 326.948 C 153.640 326.866,153.909 326.080,153.798 325.200 C 153.634 323.906,153.742 323.829,154.364 324.793 C 155.406 326.410,157.264 324.663,156.283 322.990 C 155.693 321.984,155.420 322.000,154.278 323.113 M21.641 353.124 C 21.030 354.265,20.742 355.409,21.001 355.667 C 21.613 356.279,23.388 352.963,23.027 351.881 C 22.875 351.424,22.251 351.984,21.641 353.124 " stroke="none" fill="#000000" fill-rule="evenodd"></path><path id="path1" d="M0.000 18.000 C 0.000 29.467,0.290 36.000,0.800 36.000 C 1.267 36.000,1.600 38.675,1.600 42.425 C 1.600 47.398,1.871 48.954,2.800 49.311 C 3.694 49.654,4.000 51.077,4.000 54.886 C 4.000 58.898,4.258 60.000,5.200 60.000 C 6.104 60.000,6.400 60.971,6.400 63.935 C 6.400 68.616,6.975 70.400,8.486 70.400 C 9.290 70.400,9.600 71.514,9.600 74.400 C 9.600 77.355,9.902 78.400,10.756 78.400 C 11.568 78.400,11.985 79.532,12.156 82.200 C 12.318 84.716,12.738 85.932,13.400 85.800 C 14.325 85.615,14.763 86.634,14.467 88.283 C 14.208 89.730,15.342 91.959,16.249 91.784 C 16.836 91.670,17.295 92.918,17.448 95.044 C 17.598 97.118,18.154 98.663,18.848 98.929 C 19.482 99.172,20.000 99.726,20.000 100.159 C 20.000 101.383,23.235 105.200,24.272 105.200 C 24.783 105.200,25.313 106.370,25.451 107.800 C 25.606 109.410,26.139 110.400,26.851 110.400 C 27.631 110.400,28.000 111.299,28.000 113.200 C 28.000 115.156,28.362 116.000,29.200 116.000 C 29.946 116.000,30.400 116.755,30.400 117.993 C 30.400 119.120,31.095 120.421,32.000 120.986 C 32.880 121.535,33.600 122.810,33.600 123.818 C 33.600 124.826,34.140 125.857,34.800 126.111 C 35.460 126.364,36.000 127.073,36.000 127.686 C 36.000 128.299,36.337 128.800,36.748 128.800 C 37.160 128.800,37.944 130.038,38.492 131.552 C 39.039 133.066,39.851 134.236,40.296 134.152 C 40.741 134.068,41.254 134.720,41.435 135.600 C 41.858 137.649,43.936 140.000,45.324 140.000 C 45.916 140.000,46.400 140.540,46.400 141.200 C 46.400 141.867,47.111 142.400,48.000 142.400 C 49.300 142.400,49.600 142.933,49.600 145.247 C 49.600 147.061,49.910 147.903,50.454 147.567 C 50.924 147.276,51.505 147.794,51.747 148.720 C 51.989 149.644,52.732 150.400,53.399 150.400 C 54.348 150.400,54.519 151.084,54.188 153.550 C 53.714 157.088,54.153 158.000,55.711 156.708 C 56.518 156.038,57.200 156.351,58.585 158.030 C 59.583 159.240,62.147 162.022,64.281 164.212 C 66.416 166.403,67.676 167.937,67.081 167.621 C 66.487 167.306,67.661 168.676,69.692 170.666 C 71.723 172.656,73.610 174.873,73.887 175.593 C 74.163 176.313,74.751 176.677,75.194 176.403 C 75.637 176.130,76.000 176.437,76.000 177.086 C 76.000 177.735,76.432 178.411,76.961 178.587 C 77.489 178.763,77.939 179.933,77.961 181.187 C 77.982 182.441,78.507 183.636,79.127 183.843 C 79.891 184.097,80.098 183.809,79.768 182.950 C 79.438 182.090,79.635 181.817,80.379 182.103 C 80.982 182.334,81.666 183.018,81.897 183.621 C 82.183 184.365,81.910 184.562,81.050 184.232 C 80.190 183.902,79.904 184.108,80.161 184.873 C 80.370 185.493,81.288 186.000,82.200 186.000 C 83.112 186.000,84.760 186.781,85.861 187.735 C 88.647 190.150,90.875 191.247,91.623 190.572 C 91.971 190.257,92.011 190.428,91.711 190.952 C 90.992 192.206,95.664 197.485,98.014 198.075 C 99.023 198.328,100.039 199.261,100.271 200.148 C 100.503 201.035,101.077 201.524,101.546 201.233 C 102.016 200.943,102.400 201.243,102.400 201.899 C 102.400 202.555,103.120 203.321,104.000 203.600 C 104.880 203.879,105.600 204.585,105.600 205.167 C 105.600 205.750,106.515 206.456,107.633 206.737 C 108.751 207.017,109.444 207.605,109.173 208.043 C 108.903 208.481,108.978 208.792,109.341 208.733 C 110.870 208.487,113.600 209.871,113.600 210.893 C 113.600 211.502,114.320 212.000,115.200 212.000 C 116.080 212.000,116.710 212.450,116.600 213.000 C 116.480 213.599,117.444 214.101,119.000 214.251 C 120.952 214.439,121.600 214.887,121.600 216.051 C 121.600 216.903,122.105 217.600,122.722 217.600 C 123.339 217.600,124.220 218.680,124.680 220.000 C 125.469 222.262,125.761 222.399,129.758 222.388 C 134.153 222.376,134.490 222.244,141.978 217.600 C 146.911 214.541,151.290 211.926,154.400 210.183 C 178.240 196.818,185.032 192.113,184.658 189.218 C 184.543 188.328,184.764 187.094,185.148 186.476 C 185.636 185.693,185.557 185.532,184.889 185.945 C 183.766 186.639,181.414 184.949,182.056 183.909 C 182.297 183.519,181.753 183.200,180.847 183.200 C 179.904 183.200,179.200 182.678,179.200 181.980 C 179.200 181.264,178.208 180.562,176.800 180.280 C 175.480 180.016,174.400 179.305,174.400 178.700 C 174.400 178.095,173.860 177.600,173.200 177.600 C 172.540 177.600,172.000 177.084,172.000 176.454 C 172.000 175.824,171.313 175.090,170.474 174.823 C 169.635 174.557,168.746 173.813,168.499 173.170 C 168.252 172.526,167.139 172.000,166.025 172.000 C 164.758 172.000,164.000 171.551,164.000 170.800 C 164.000 170.140,163.502 169.600,162.893 169.600 C 162.285 169.600,161.598 168.880,161.368 168.000 C 161.138 167.120,160.376 166.400,159.675 166.400 C 158.853 166.400,158.400 165.699,158.400 164.428 C 158.400 163.248,157.730 162.098,156.731 161.563 C 155.814 161.072,155.236 160.497,155.448 160.286 C 155.999 159.734,152.393 156.000,151.308 156.000 C 150.809 156.000,150.400 155.460,150.400 154.800 C 150.400 154.140,149.860 153.600,149.200 153.600 C 148.540 153.600,148.000 153.214,148.000 152.743 C 148.000 151.559,145.902 149.564,145.295 150.171 C 145.023 150.444,144.800 150.067,144.800 149.333 C 144.800 148.600,144.260 148.000,143.600 148.000 C 142.940 148.000,142.400 147.460,142.400 146.800 C 142.400 146.140,141.860 145.600,141.200 145.600 C 140.403 145.600,140.000 144.803,140.000 143.228 C 140.000 141.664,139.447 140.561,138.378 139.988 C 137.485 139.510,136.967 138.778,137.225 138.360 C 137.483 137.942,136.953 137.600,136.047 137.600 C 134.698 137.600,134.400 137.093,134.400 134.800 C 134.400 132.844,134.038 132.000,133.200 132.000 C 132.540 132.000,132.000 131.460,132.000 130.800 C 132.000 130.140,131.267 128.867,130.371 127.971 C 129.475 127.075,128.971 125.971,129.251 125.517 C 129.532 125.063,129.005 124.495,128.080 124.253 C 126.841 123.929,126.400 123.179,126.400 121.395 C 126.400 120.065,125.875 118.577,125.234 118.088 C 124.593 117.600,123.881 116.121,123.651 114.802 C 123.422 113.483,122.596 111.927,121.817 111.345 C 121.038 110.763,120.505 109.907,120.633 109.443 C 120.945 108.317,118.430 105.600,117.076 105.600 C 116.318 105.600,116.000 104.417,116.000 101.600 C 116.000 98.578,115.707 97.600,114.800 97.600 C 113.762 97.600,113.112 95.090,113.505 92.600 C 113.557 92.270,113.219 92.000,112.753 92.000 C 112.287 92.000,112.130 91.637,112.403 91.194 C 112.677 90.751,112.339 90.173,111.651 89.909 C 110.752 89.564,110.400 88.326,110.400 85.514 C 110.400 82.568,110.103 81.600,109.200 81.600 C 108.111 81.600,107.404 76.328,107.940 72.200 C 108.012 71.650,107.514 71.200,106.835 71.200 C 105.923 71.200,105.587 70.310,105.549 67.800 C 105.454 61.469,104.966 58.900,103.681 57.960 C 102.653 57.209,102.400 55.432,102.400 48.969 C 102.400 42.700,102.120 40.635,101.138 39.652 C 100.059 38.574,99.873 35.595,99.859 19.195 L 99.842 0.000 49.921 0.000 L 0.000 0.000 0.000 18.000 M258.000 241.418 C 257.560 241.843,254.500 243.646,251.200 245.425 C 247.900 247.203,240.880 251.190,235.600 254.285 C 230.320 257.380,223.818 261.185,221.152 262.740 C 212.569 267.746,205.900 276.000,210.438 276.000 C 212.709 276.000,217.559 278.852,217.579 280.200 C 217.591 280.970,218.050 281.615,218.600 281.634 C 219.150 281.653,220.483 282.733,221.562 284.034 C 222.642 285.335,223.992 286.400,224.562 286.400 C 225.133 286.400,225.600 286.686,225.600 287.035 C 225.600 288.030,231.774 294.400,232.737 294.400 C 233.212 294.400,233.600 294.898,233.600 295.507 C 233.600 296.115,234.330 296.804,235.223 297.038 C 236.116 297.271,236.746 297.943,236.623 298.531 C 236.500 299.119,236.828 299.517,237.351 299.416 C 237.874 299.315,238.529 299.945,238.805 300.816 C 239.082 301.687,239.824 302.400,240.454 302.400 C 241.775 302.400,241.985 303.667,240.787 304.408 C 239.581 305.153,240.768 306.583,242.151 306.053 C 242.867 305.778,244.181 306.554,245.449 307.998 C 246.608 309.319,248.016 310.400,248.578 310.400 C 249.185 310.400,249.600 311.395,249.600 312.847 C 249.600 314.226,249.939 315.085,250.377 314.814 C 250.804 314.551,251.383 315.249,251.663 316.367 C 251.944 317.485,252.466 318.400,252.823 318.400 C 254.193 318.400,259.683 324.663,261.456 328.247 C 262.477 330.311,263.827 332.000,264.456 332.000 C 265.127 332.000,265.600 332.786,265.600 333.902 C 265.600 334.948,266.140 336.252,266.800 336.800 C 267.460 337.348,268.000 338.663,268.000 339.723 C 268.000 340.783,268.540 341.857,269.200 342.111 C 269.860 342.364,270.411 343.163,270.425 343.886 C 270.458 345.550,272.388 348.146,273.079 347.454 C 273.366 347.168,273.600 348.253,273.600 349.867 C 273.600 351.941,273.952 352.800,274.800 352.800 C 275.805 352.800,276.216 354.109,275.947 356.461 C 275.919 356.714,276.694 357.614,277.670 358.461 C 278.646 359.307,279.166 360.000,278.824 360.000 C 278.483 360.000,278.605 360.483,279.095 361.074 C 279.585 361.664,279.719 362.441,279.393 362.801 C 279.067 363.161,279.430 363.229,280.200 362.953 C 281.142 362.615,281.600 362.850,281.600 363.672 C 281.600 364.421,281.213 364.673,280.600 364.323 C 280.050 364.009,280.590 364.818,281.800 366.120 C 283.376 367.816,284.000 369.326,284.000 371.444 C 284.000 373.538,284.350 374.400,285.200 374.400 C 286.308 374.400,286.895 377.352,286.445 380.662 C 286.366 381.246,287.044 381.918,287.951 382.155 C 289.310 382.511,289.600 383.223,289.600 386.205 C 289.600 388.681,290.042 390.186,291.000 390.971 C 291.856 391.673,291.934 391.895,291.200 391.542 C 290.226 391.074,290.188 391.168,291.000 392.043 C 291.578 392.666,292.000 395.228,292.000 398.120 C 292.000 401.640,292.341 403.311,293.152 403.765 C 293.904 404.186,294.353 405.933,294.447 408.805 C 294.641 414.752,295.164 416.800,296.487 416.800 C 297.365 416.800,297.613 418.616,297.661 425.400 C 297.731 435.210,297.934 436.818,299.202 437.601 C 299.841 437.996,300.112 444.344,300.116 459.081 L 300.123 480.000 350.061 480.000 L 400.000 480.000 400.000 460.800 C 400.000 448.533,399.711 441.600,399.200 441.600 C 398.723 441.600,398.400 438.525,398.400 433.975 C 398.400 427.935,398.151 426.254,397.200 425.889 C 396.303 425.545,396.000 424.110,396.000 420.203 C 396.000 416.560,395.646 414.707,394.832 414.088 C 394.112 413.540,393.595 411.548,393.484 408.886 C 393.342 405.484,392.997 404.492,391.852 404.193 C 390.656 403.880,390.400 402.980,390.400 399.082 C 390.400 395.614,390.080 394.227,389.200 393.889 C 388.473 393.610,388.000 392.401,388.000 390.822 C 388.000 389.389,387.257 387.271,386.348 386.116 C 385.439 384.961,384.942 383.618,385.242 383.132 C 385.543 382.646,385.023 381.212,384.087 379.947 C 383.151 378.681,382.304 376.654,382.204 375.442 C 382.102 374.221,380.858 372.041,379.411 370.551 C 377.975 369.072,376.800 367.342,376.800 366.706 C 376.800 366.070,376.260 365.343,375.600 365.089 C 374.940 364.836,374.400 363.793,374.400 362.772 C 374.400 361.750,373.860 360.374,373.200 359.714 C 372.540 359.054,372.000 357.606,372.000 356.497 C 372.000 355.387,371.453 354.173,370.785 353.799 C 370.116 353.425,369.401 352.238,369.194 351.160 C 368.988 350.082,368.334 348.840,367.742 348.400 C 367.150 347.960,366.419 346.610,366.117 345.400 C 365.816 344.190,365.216 343.200,364.785 343.200 C 364.353 343.200,364.000 342.522,364.000 341.693 C 364.000 340.858,363.285 340.000,362.397 339.767 C 361.429 339.514,360.965 338.902,361.225 338.224 C 361.462 337.605,360.924 336.436,360.028 335.626 C 359.133 334.815,358.400 333.560,358.400 332.838 C 358.400 332.115,357.151 330.280,355.625 328.761 C 354.099 327.243,352.839 325.529,352.825 324.953 C 352.811 324.377,352.416 324.143,351.946 324.433 C 351.476 324.724,350.895 324.206,350.653 323.280 C 350.411 322.356,349.715 321.600,349.107 321.600 C 348.498 321.600,348.000 320.880,348.000 320.000 C 348.000 319.111,347.467 318.400,346.800 318.400 C 346.042 318.400,345.600 317.636,345.600 316.325 C 345.600 314.923,345.081 314.115,344.000 313.832 C 343.114 313.600,342.400 312.741,342.400 311.907 C 342.400 311.078,341.959 310.490,341.419 310.600 C 340.879 310.710,340.314 309.990,340.163 309.000 C 340.012 308.010,339.494 307.200,339.011 307.200 C 338.528 307.200,338.359 306.975,338.634 306.699 C 338.910 306.424,338.070 305.264,336.767 304.121 C 335.465 302.979,334.400 301.563,334.400 300.975 C 334.400 300.387,334.016 300.143,333.546 300.433 C 333.076 300.724,332.495 300.206,332.253 299.280 C 331.997 298.301,331.165 297.600,330.259 297.600 C 329.405 297.600,328.903 297.281,329.144 296.891 C 329.688 296.010,327.650 294.217,326.926 294.940 C 326.637 295.230,326.400 294.687,326.400 293.733 C 326.400 292.726,325.897 292.000,325.200 292.000 C 324.540 292.000,324.000 291.460,324.000 290.800 C 324.000 290.140,323.460 289.600,322.800 289.600 C 322.140 289.600,321.600 289.026,321.600 288.325 C 321.600 287.624,320.880 286.862,320.000 286.632 C 319.120 286.402,318.400 285.715,318.400 285.107 C 318.400 284.498,318.021 284.000,317.558 284.000 C 317.095 284.000,316.340 282.920,315.880 281.600 C 315.420 280.280,314.719 279.200,314.322 279.200 C 313.925 279.200,312.880 278.480,312.000 277.600 C 311.120 276.720,309.860 276.000,309.200 276.000 C 308.540 276.000,308.000 275.502,308.000 274.893 C 308.000 274.285,307.270 273.596,306.377 273.362 C 305.484 273.129,304.854 272.473,304.977 271.905 C 305.100 271.337,304.840 270.856,304.400 270.836 C 303.342 270.788,300.000 267.419,300.000 266.400 C 300.000 265.960,299.503 265.600,298.895 265.600 C 298.288 265.600,296.230 263.800,294.321 261.600 C 292.413 259.400,290.390 257.600,289.826 257.600 C 289.262 257.600,288.800 257.060,288.800 256.400 C 288.800 255.338,287.288 254.784,285.331 255.128 C 284.963 255.193,284.471 254.516,284.238 253.623 C 284.004 252.730,283.304 252.000,282.682 252.000 C 282.059 252.000,281.343 251.460,281.089 250.800 C 280.827 250.116,279.634 249.600,278.314 249.600 C 276.533 249.600,276.000 249.231,276.000 248.000 C 276.000 246.918,275.467 246.400,274.353 246.400 C 273.447 246.400,272.930 246.037,273.203 245.594 C 273.477 245.151,273.049 244.544,272.251 244.244 C 271.453 243.944,270.440 243.221,270.000 242.637 C 269.020 241.337,259.123 240.331,258.000 241.418 " stroke="none" fill="#9bc848" fill-rule="evenodd"></path><path id="path2" d="M300.000 16.857 C 300.000 30.775,299.791 33.924,298.800 34.914 C 297.924 35.790,297.600 37.926,297.600 42.817 C 297.600 48.170,297.338 49.667,296.301 50.247 C 295.360 50.774,294.889 52.564,294.589 56.755 C 294.315 60.581,293.807 62.678,293.088 62.954 C 292.376 63.227,292.000 64.724,292.000 67.286 C 292.000 70.232,291.703 71.200,290.800 71.200 C 289.921 71.200,289.600 72.112,289.600 74.608 C 289.600 77.149,289.193 78.269,288.000 79.014 C 287.076 79.591,286.400 80.880,286.400 82.064 C 286.400 83.192,285.860 84.654,285.200 85.314 C 284.540 85.974,284.000 87.551,284.000 88.817 C 284.000 90.084,283.517 91.408,282.926 91.760 C 282.335 92.112,281.171 93.480,280.340 94.800 C 279.508 96.120,278.192 97.670,277.414 98.245 C 276.636 98.820,276.000 100.170,276.000 101.245 C 276.000 102.449,275.539 103.200,274.800 103.200 C 274.000 103.200,273.600 104.000,273.600 105.600 C 273.600 107.068,273.185 108.000,272.532 108.000 C 271.945 108.000,270.595 109.786,269.532 111.969 C 268.469 114.152,267.150 115.952,266.600 115.969 C 265.996 115.988,265.600 117.110,265.600 118.800 C 265.600 121.067,265.295 121.600,264.000 121.600 C 263.111 121.600,262.400 122.133,262.400 122.800 C 262.400 123.460,261.860 124.000,261.200 124.000 C 260.540 124.000,260.000 124.501,260.000 125.114 C 260.000 125.727,259.460 126.436,258.800 126.689 C 258.140 126.943,257.600 127.701,257.600 128.375 C 257.600 129.049,257.240 129.600,256.800 129.600 C 255.758 129.600,249.600 135.758,249.600 136.800 C 249.600 137.240,249.060 137.600,248.400 137.600 C 247.740 137.600,247.200 137.929,247.200 138.331 C 247.200 138.734,245.940 140.360,244.400 141.946 C 242.860 143.532,241.600 145.122,241.600 145.479 C 241.600 145.837,240.610 146.635,239.400 147.254 C 236.456 148.757,233.631 151.368,233.614 152.600 C 233.606 153.150,233.049 153.600,232.375 153.600 C 231.701 153.600,230.943 154.140,230.689 154.800 C 230.436 155.460,229.550 156.000,228.721 156.000 C 227.892 156.000,226.764 156.720,226.214 157.600 C 225.665 158.480,224.762 159.200,224.208 159.200 C 223.653 159.200,223.200 159.740,223.200 160.400 C 223.200 161.060,222.649 161.600,221.975 161.600 C 221.301 161.600,220.543 162.140,220.289 162.800 C 220.036 163.460,219.327 164.000,218.714 164.000 C 218.101 164.000,217.600 164.498,217.600 165.107 C 217.600 165.715,216.880 166.402,216.000 166.632 C 215.120 166.862,214.400 167.585,214.400 168.238 C 214.400 168.892,213.500 169.652,212.400 169.929 C 211.300 170.205,210.400 170.784,210.400 171.215 C 210.400 171.647,209.669 172.000,208.775 172.000 C 207.881 172.000,206.951 172.518,206.708 173.151 C 206.465 173.784,205.216 174.533,203.933 174.815 C 202.650 175.097,201.600 175.839,201.600 176.464 C 201.600 177.089,201.049 177.600,200.375 177.600 C 199.701 177.600,198.943 178.140,198.689 178.800 C 198.436 179.460,197.356 180.000,196.289 180.000 C 195.223 180.000,194.155 180.508,193.917 181.128 C 193.679 181.748,192.543 182.492,191.392 182.781 C 190.241 183.070,188.889 183.867,188.387 184.553 C 187.886 185.239,186.475 185.800,185.252 185.800 C 184.030 185.800,183.339 186.025,183.717 186.300 C 188.367 189.678,182.604 194.372,154.400 210.183 C 151.290 211.926,146.911 214.541,141.978 217.600 C 134.569 222.195,134.139 222.376,130.600 222.388 C 128.278 222.396,126.693 222.907,125.600 224.000 C 124.720 224.880,123.460 225.600,122.800 225.600 C 122.140 225.600,121.600 226.140,121.600 226.800 C 121.600 227.460,121.049 228.000,120.375 228.000 C 119.701 228.000,118.951 228.518,118.708 229.151 C 118.465 229.784,117.216 230.533,115.933 230.815 C 114.650 231.097,113.600 231.839,113.600 232.464 C 113.600 233.089,113.060 233.600,112.400 233.600 C 111.740 233.600,111.200 234.140,111.200 234.800 C 111.200 235.460,110.750 236.006,110.200 236.014 C 108.980 236.031,105.600 239.484,105.600 240.713 C 105.600 241.201,105.049 241.600,104.375 241.600 C 103.701 241.600,102.943 242.140,102.689 242.800 C 102.436 243.460,101.727 244.000,101.114 244.000 C 100.501 244.000,100.000 244.540,100.000 245.200 C 100.000 245.860,99.460 246.400,98.800 246.400 C 98.133 246.400,97.600 247.111,97.600 248.000 C 97.600 249.067,97.067 249.600,96.000 249.600 C 95.111 249.600,94.400 250.133,94.400 250.800 C 94.400 251.460,93.860 252.000,93.200 252.000 C 92.540 252.000,91.280 252.720,90.400 253.600 C 89.520 254.480,88.440 255.200,88.000 255.200 C 87.560 255.200,87.200 255.653,87.200 256.208 C 87.200 256.762,86.480 257.665,85.600 258.214 C 84.720 258.764,84.000 259.892,84.000 260.721 C 84.000 261.550,83.460 262.436,82.800 262.689 C 82.140 262.943,81.600 263.701,81.600 264.375 C 81.600 265.049,81.049 265.600,80.375 265.600 C 79.701 265.600,78.943 266.140,78.689 266.800 C 78.436 267.460,77.727 268.000,77.114 268.000 C 76.501 268.000,76.000 268.501,76.000 269.114 C 76.000 269.727,75.460 270.436,74.800 270.689 C 74.140 270.943,73.600 271.701,73.600 272.375 C 73.600 273.049,73.060 273.600,72.400 273.600 C 71.740 273.600,71.200 273.971,71.200 274.425 C 71.200 274.878,69.940 276.406,68.400 277.819 C 66.860 279.233,65.600 280.661,65.600 280.994 C 65.600 281.328,64.970 282.320,64.200 283.200 C 63.430 284.080,62.170 285.520,61.400 286.400 C 60.630 287.280,60.000 288.321,60.000 288.714 C 60.000 289.107,59.460 289.636,58.800 289.889 C 58.140 290.143,57.600 291.050,57.600 291.906 C 57.600 292.762,56.340 294.760,54.800 296.346 C 53.260 297.932,52.000 299.904,52.000 300.729 C 52.000 301.554,51.460 302.436,50.800 302.689 C 50.099 302.958,49.600 304.157,49.600 305.575 C 49.600 307.055,49.187 308.000,48.539 308.000 C 47.295 308.000,44.000 311.314,44.000 312.566 C 44.000 313.041,43.460 313.636,42.800 313.889 C 42.140 314.143,41.600 315.063,41.600 315.935 C 41.600 316.807,41.150 317.808,40.600 318.160 C 39.458 318.891,37.990 321.091,36.704 324.000 C 36.218 325.100,35.320 326.956,34.710 328.123 C 34.099 329.291,33.600 331.139,33.600 332.230 C 33.600 333.541,33.057 334.355,32.000 334.632 C 30.835 334.936,30.400 335.723,30.400 337.525 C 30.400 339.192,30.008 340.000,29.200 340.000 C 28.362 340.000,28.000 340.844,28.000 342.800 C 28.000 344.756,27.638 345.600,26.800 345.600 C 25.962 345.600,25.600 346.444,25.600 348.400 C 25.600 350.096,25.206 351.208,24.600 351.221 C 24.050 351.232,22.913 352.762,22.074 354.621 C 21.235 356.479,19.885 358.644,19.074 359.431 C 18.131 360.347,17.600 362.004,17.600 364.031 C 17.600 366.255,17.257 367.200,16.449 367.200 C 15.690 367.200,15.105 368.486,14.732 370.974 C 14.421 373.050,13.679 375.236,13.083 375.831 C 12.438 376.476,12.000 378.672,12.000 381.257 C 12.000 384.584,11.719 385.600,10.800 385.600 C 9.906 385.600,9.600 386.546,9.600 389.312 C 9.600 391.878,9.181 393.330,8.244 394.015 C 7.268 394.729,6.793 396.546,6.548 400.503 C 6.329 404.032,5.812 406.229,5.104 406.640 C 4.321 407.094,4.000 409.020,4.000 413.265 C 4.000 417.847,3.719 419.358,2.800 419.711 C 1.845 420.077,1.600 421.789,1.600 428.086 C 1.600 432.829,1.279 436.000,0.800 436.000 C 0.286 436.000,0.000 443.867,0.000 458.000 L 0.000 480.000 49.507 480.000 L 99.013 480.000 99.496 478.200 C 99.761 477.210,99.983 468.794,99.989 459.498 C 99.998 445.459,100.203 442.427,101.200 441.600 C 102.113 440.842,102.400 438.782,102.400 432.990 C 102.400 426.934,102.662 425.185,103.680 424.441 C 104.551 423.803,105.098 421.732,105.392 417.952 C 105.660 414.512,106.238 412.157,106.912 411.760 C 107.637 411.334,108.000 409.522,108.000 406.335 C 108.000 402.819,108.318 401.428,109.200 401.089 C 110.001 400.782,110.400 399.520,110.400 397.292 C 110.400 394.946,110.875 393.549,112.000 392.586 C 113.041 391.694,113.600 390.226,113.600 388.383 C 113.600 386.603,114.046 385.379,114.800 385.089 C 115.467 384.833,116.000 383.645,116.000 382.412 C 116.000 381.193,116.540 379.748,117.200 379.200 C 117.860 378.652,118.400 377.348,118.400 376.302 C 118.400 375.145,118.870 374.400,119.600 374.400 C 120.260 374.400,120.800 373.940,120.800 373.378 C 120.800 372.816,121.520 371.739,122.400 370.986 C 123.280 370.232,124.000 368.936,124.000 368.106 C 124.000 367.275,124.540 366.148,125.200 365.600 C 125.860 365.052,126.400 363.748,126.400 362.702 C 126.400 361.656,126.755 360.800,127.189 360.800 C 127.623 360.800,128.487 359.584,129.108 358.098 C 129.729 356.611,130.633 355.150,131.118 354.850 C 131.603 354.551,132.000 353.427,132.000 352.353 C 132.000 351.150,132.461 350.400,133.200 350.400 C 133.989 350.400,134.400 349.610,134.400 348.093 C 134.400 346.459,134.867 345.665,136.000 345.368 C 136.880 345.138,137.600 344.415,137.600 343.760 C 137.600 343.106,138.140 342.364,138.800 342.111 C 139.460 341.857,140.000 340.739,140.000 339.625 C 140.000 338.511,140.361 337.600,140.802 337.600 C 141.927 337.600,148.000 331.023,148.000 329.804 C 148.000 329.252,148.540 328.800,149.200 328.800 C 149.860 328.800,150.400 328.260,150.400 327.600 C 150.400 326.940,150.940 326.400,151.600 326.400 C 152.260 326.400,152.800 325.878,152.800 325.240 C 152.800 324.602,153.250 323.792,153.800 323.440 C 155.125 322.592,158.400 317.888,158.400 316.832 C 158.400 316.375,158.850 315.994,159.400 315.986 C 160.620 315.969,164.000 312.516,164.000 311.287 C 164.000 310.799,164.540 310.400,165.200 310.400 C 165.860 310.400,166.400 309.860,166.400 309.200 C 166.400 308.540,166.850 307.989,167.400 307.975 C 167.950 307.961,169.643 306.701,171.161 305.175 C 172.680 303.649,174.372 302.400,174.922 302.400 C 175.471 302.400,176.208 301.923,176.560 301.340 C 176.912 300.756,177.830 300.048,178.600 299.765 C 179.370 299.482,180.000 298.699,180.000 298.025 C 180.000 297.351,180.540 296.800,181.200 296.800 C 181.860 296.800,182.400 296.260,182.400 295.600 C 182.400 294.940,182.940 294.400,183.600 294.400 C 184.260 294.400,184.800 293.860,184.800 293.200 C 184.800 292.540,185.250 291.988,185.800 291.973 C 187.356 291.932,190.400 289.086,190.400 287.672 C 190.400 286.802,191.132 286.400,192.714 286.400 C 194.034 286.400,195.227 285.884,195.489 285.200 C 195.743 284.540,196.501 284.000,197.175 284.000 C 197.849 284.000,198.400 283.460,198.400 282.800 C 198.400 282.140,198.940 281.600,199.600 281.600 C 200.260 281.600,201.520 280.880,202.400 280.000 C 203.280 279.120,204.540 278.400,205.200 278.400 C 205.860 278.400,206.400 277.860,206.400 277.200 C 206.400 276.540,207.080 276.000,207.911 276.000 C 209.003 276.000,209.567 275.227,209.945 273.210 C 210.726 269.046,213.445 267.065,237.200 253.352 C 242.920 250.050,249.760 246.183,252.400 244.759 C 255.040 243.335,257.652 241.816,258.204 241.383 C 258.756 240.951,261.053 240.774,263.307 240.991 C 266.824 241.329,267.461 241.173,267.796 239.892 C 268.011 239.072,268.685 238.400,269.293 238.400 C 269.902 238.400,270.400 237.860,270.400 237.200 C 270.400 236.428,271.184 235.995,272.600 235.986 C 274.782 235.972,278.400 233.041,278.400 231.287 C 278.400 230.799,279.491 230.400,280.825 230.400 C 282.243 230.400,283.442 229.901,283.711 229.200 C 283.964 228.540,284.673 228.000,285.286 228.000 C 285.899 228.000,286.400 227.460,286.400 226.800 C 286.400 226.140,286.940 225.600,287.600 225.600 C 288.260 225.600,289.458 224.942,290.261 224.139 C 291.065 223.335,292.325 222.447,293.061 222.164 C 293.798 221.881,294.400 221.099,294.400 220.425 C 294.400 219.751,294.700 219.200,295.067 219.200 C 295.433 219.200,296.783 218.130,298.067 216.822 C 299.350 215.514,300.850 214.434,301.400 214.422 C 301.950 214.410,302.400 213.860,302.400 213.200 C 302.400 212.540,302.741 212.000,303.159 212.000 C 304.261 212.000,308.000 208.376,308.000 207.308 C 308.000 206.809,308.540 206.400,309.200 206.400 C 309.860 206.400,310.400 205.860,310.400 205.200 C 310.400 204.540,311.090 204.000,311.933 204.000 C 313.461 204.000,318.400 198.566,318.400 196.885 C 318.400 196.398,318.799 196.000,319.287 196.000 C 320.516 196.000,323.969 192.620,323.986 191.400 C 323.994 190.850,324.540 190.400,325.200 190.400 C 325.860 190.400,326.400 190.028,326.400 189.572 C 326.400 189.117,328.740 186.416,331.600 183.570 C 334.460 180.724,336.800 178.217,336.800 177.998 C 336.800 177.779,337.520 176.880,338.400 176.000 C 339.280 175.120,340.000 173.860,340.000 173.200 C 340.000 172.540,340.540 172.000,341.200 172.000 C 341.860 172.000,342.400 171.460,342.400 170.800 C 342.400 170.140,343.120 168.880,344.000 168.000 C 344.880 167.120,345.600 165.860,345.600 165.200 C 345.600 164.540,346.140 164.000,346.800 164.000 C 347.467 164.000,348.000 163.289,348.000 162.400 C 348.000 161.511,348.533 160.800,349.200 160.800 C 349.989 160.800,350.400 160.010,350.400 158.493 C 350.400 156.859,350.867 156.065,352.000 155.768 C 352.880 155.538,353.600 154.776,353.600 154.075 C 353.600 153.374,354.140 152.800,354.800 152.800 C 355.460 152.800,356.000 152.299,356.000 151.686 C 356.000 151.073,356.540 150.364,357.200 150.111 C 357.860 149.857,358.400 149.099,358.400 148.425 C 358.400 147.751,358.779 147.200,359.242 147.200 C 359.705 147.200,360.450 146.149,360.898 144.864 C 361.345 143.579,362.227 142.331,362.856 142.089 C 363.485 141.848,364.000 140.728,364.000 139.600 C 364.000 138.472,364.540 137.343,365.200 137.089 C 365.884 136.827,366.400 135.634,366.400 134.314 C 366.400 132.869,366.819 132.000,367.514 132.000 C 368.738 132.000,369.600 130.319,369.600 127.935 C 369.600 127.091,370.140 126.400,370.800 126.400 C 371.603 126.400,372.000 125.597,372.000 123.975 C 372.000 122.557,372.499 121.358,373.200 121.089 C 373.884 120.827,374.400 119.634,374.400 118.314 C 374.400 116.848,374.816 116.000,375.536 116.000 C 376.161 116.000,376.913 114.908,377.206 113.573 C 377.499 112.239,378.788 109.999,380.069 108.595 C 381.425 107.111,382.400 105.149,382.400 103.907 C 382.400 102.732,382.920 101.572,383.556 101.328 C 384.192 101.084,384.967 99.524,385.279 97.862 C 385.590 96.200,386.330 94.355,386.923 93.763 C 387.515 93.170,388.000 91.310,388.000 89.628 C 388.000 87.649,388.423 86.409,389.200 86.111 C 390.083 85.772,390.400 84.376,390.400 80.825 C 390.400 77.234,390.685 76.000,391.514 76.000 C 392.982 76.000,393.600 74.232,393.600 70.033 C 393.600 67.954,394.074 66.202,394.800 65.600 C 395.637 64.905,396.000 63.061,396.000 59.502 C 396.000 55.500,396.259 54.400,397.200 54.400 C 398.187 54.400,398.400 52.978,398.400 46.400 C 398.400 41.600,398.720 38.400,399.200 38.400 C 399.711 38.400,400.000 31.467,400.000 19.200 L 400.000 0.000 350.000 0.000 L 300.000 0.000 300.000 16.857 M99.418 18.400 C 99.418 20.380,99.574 21.190,99.765 20.200 C 99.955 19.210,99.955 17.590,99.765 16.600 C 99.574 15.610,99.418 16.420,99.418 18.400 M300.202 459.600 C 300.211 461.360,300.374 461.986,300.566 460.991 C 300.757 459.996,300.750 458.556,300.550 457.791 C 300.349 457.026,300.193 457.840,300.202 459.600 " stroke="none" fill="#1982ae" fill-rule="evenodd"></path></g></svg>


					
					</a></li>
			
				<li><a target="_blank" href="https://plus.google.com/+Dnanexus" class="Google Plus-icon">
					
						
		<svg class="google-plus" fill="#000000" height="24" viewBox="0 0 24 24" width="24" xmlns="http://www.w3.org/2000/svg"><path d="M23,11H21V9H19V11H17V13H19V15H21V13H23M8,11V13.4H12C11.8,14.4 10.8,16.4 8,16.4C5.6,16.4 3.7,14.4 3.7,12C3.7,9.6 5.6,7.6 8,7.6C9.4,7.6 10.3,8.2 10.8,8.7L12.7,6.9C11.5,5.7 9.9,5 8,5C4.1,5 1,8.1 1,12C1,15.9 4.1,19 8,19C12,19 14.7,16.2 14.7,12.2C14.7,11.7 14.7,11.4 14.6,11H8Z" /></svg>
	

					
					</a></li>
			
				<li><a target="_blank" href="https://www.linkedin.com/company-beta/912957/" class="LinkedIn-icon">
					
						
		<svg class="linkedin" fill="#000000" height="24" viewBox="0 0 24 24" width="24" xmlns="http://www.w3.org/2000/svg"><path d="M19,19H16V13.7A1.5,1.5 0 0,0 14.5,12.2A1.5,1.5 0 0,0 13,13.7V19H10V10H13V11.2C13.5,10.36 14.59,9.8 15.5,9.8A3.5,3.5 0 0,1 19,13.3M6.5,8.31C5.5,8.31 4.69,7.5 4.69,6.5A1.81,1.81 0 0,1 6.5,4.69C7.5,4.69 8.31,5.5 8.31,6.5A1.81,1.81 0 0,1 6.5,8.31M8,19H5V10H8M20,2H4C2.89,2 2,2.89 2,4V20A2,2 0 0,0 4,22H20A2,2 0 0,0 22,20V4C22,2.89 21.1,2 20,2Z" /></svg>
	

					
					</a></li>
			
		</ul>
		<p class="copyright">&copy; DNAnexus Example Applets 2019. All rights reserved.</p>
	</div>
</footer>
		<script>
			$(function() {
				$('a[href*=\\#]').not(".no-smooth").on('click', function(event){
					var el = $(this.hash);
					if (el.length > 0) {
						// event.preventDefault();
						$('html,body').animate({scrollTop:$(this.hash).offset().top - 50}, 500);
					}
				});

				$('svg').click(function() {
					$(this).parent('form').submit();
				});
			});

			document.getElementById("open-nav").addEventListener("click", function (event) {
				event.preventDefault();
				document.body.classList.toggle("nav-open");
			});
		</script>
	</body>
</html>
